// @studnicky/fsm/dist/MachineRegistry.js
import { MachineAlreadyRegisteredError } from './MachineAlreadyRegisteredError.js';
export class MachineRegistry {
    static #registry = new Map();
    static register(name, interpreter) {
        if (MachineRegistry.#registry.has(name)) {
            throw new MachineAlreadyRegisteredError(name);
        }
        MachineRegistry.#registry.set(name, interpreter);
        this.onRegister(name);
    }
    static unregister(name) {
        MachineRegistry.#registry.delete(name);
        this.onUnregister(name);
    }
    static get(name) {
        const result = MachineRegistry.#registry.get(name);
        if (result === undefined) {
            this.onResolveMiss(name);
        }
        return result;
    }
    static has(name) {
        const result = MachineRegistry.#registry.has(name);
        return result;
    }
    static list() {
        const keys = MachineRegistry.#registry.keys();
        return [...keys];
    }
    // ---------------------------------------------------------------------------
    // Lifecycle hooks — no-op by default. The bare class does NO observability;
    // override in a subclass to add logging/tracing/metrics.
    // Overrides must not throw or block.
    // ---------------------------------------------------------------------------
    /** Fires after a named interpreter is successfully registered. */
    static onRegister(_id) { }
    /** Fires after a named interpreter is unregistered (whether or not it existed). */
    static onUnregister(_id) { }
    /** Fires when `get()` is called for an id not present in the registry. */
    static onResolveMiss(_id) { }
}
