// @studnicky/sample-buffer/dist/sample-buffer/SampleBuffer.js
/**
 * Fixed-capacity circular buffer for numeric samples with percentile calculation
 */
import { EMPTY_LENGTH, FIRST_ARRAY_INDEX, INCREMENT_BY_ONE, INITIAL_BUFFER_COUNT, INITIAL_BUFFER_HEAD, LAST_ARRAY_INDEX, PERCENTILE_MAX } from '../constants/index.js';
import { SampleBufferOptionsEntity } from '../entities/SampleBufferOptionsEntity.js';
import { SampleBufferError } from '../errors/index.js';
import { SampleBufferBuilder } from './SampleBufferBuilder.js';
/**
 * Fixed-capacity circular buffer for numeric samples
 *
 * Stores a sliding window of numeric measurements and calculates
 * percentiles using linear interpolation. The buffer overwrites
 * the oldest samples when full.
 *
 * Subclasses may override the protected hook methods to observe lifecycle
 * events. All hooks have no-op defaults — super() call is not required
 * unless the subclass needs base behavior.
 *
 * Fire points:
 * - `onOverflow(value)` — start of full-buffer push path, before eviction
 * - `onEvict(oldValue)` — before overwrite in the full-buffer push path
 * - `onPush(value, evicted)` — end of push(), after length/head updated
 * - `onComputeStart(length)` — start of buildSortedSamples(), before sort
 * - `onComputeComplete(length, sorted)` — end of buildSortedSamples(), after sort
 * - `onClear()` — start of clear(), before state is reset
 * - `onPercentile(pct, result)` — before returning from percentile()
 *
 * @example Basic usage
 * ```typescript
 * const buffer = SampleBuffer.create({ capacity: 100 });
 *
 * // Add samples
 * for (let i = 0; i < 100; i++) {
 *   buffer.push(Math.random() * 200);
 * }
 *
 * // Get percentiles
 * console.log(`p50: ${buffer.percentile(50)}`);
 * console.log(`p95: ${buffer.percentile(95)}`);
 * console.log(`p99: ${buffer.percentile(99)}`);
 * ```
 */
export class SampleBuffer {
    static builder() {
        const result = SampleBufferBuilder.create((options) => {
            const buffer = SampleBuffer.create(options);
            return buffer;
        });
        return result;
    }
    static create(options) {
        return new this(options);
    }
    static #validate(options) {
        if (!SampleBufferOptionsEntity.validate(options)) {
            const errors = SampleBufferOptionsEntity.validate.errors ?? [];
            const parts = errors.map((e) => {
                const part = `${e.instancePath} ${e.message}`.trim();
                return part;
            });
            const message = parts.join('; ');
            throw new SampleBufferError(message.length > 0 ? message : 'invalid options');
        }
        // Domain invariant: capacity must be a positive integer (schema enforces minimum:1 and type:integer)
    }
    _length = INITIAL_BUFFER_COUNT;
    _capacity;
    _head = INITIAL_BUFFER_HEAD;
    _samples;
    _sortedCache = null;
    /**
     * Create a new sample buffer
     *
     * @param options - Construction options including capacity
     */
    constructor(options) {
        SampleBuffer.#validate(options);
        this._capacity = options.capacity;
        this._samples = Array.from({ 'length': options.capacity });
    }
    /**
     * Clear all samples from the buffer.
     * Fire point: `onClear` is called at the start, before state is reset.
     */
    clear() {
        this.onClear();
        this._length = INITIAL_BUFFER_COUNT;
        this._head = INITIAL_BUFFER_HEAD;
        this._sortedCache = null;
    }
    /**
     * Get samples in window order, then sort them ascending.
     * Called internally by percentile(); subclasses may override to
     * customize the sort or sample extraction.
     */
    buildSortedSamples() {
        this.onComputeStart(this._length);
        const result = Array.from({ 'length': this._length });
        for (let i = FIRST_ARRAY_INDEX; i < this._length; i++) {
            const index = this._length < this._capacity
                ? i
                : (this._head + i) % this._capacity;
            result[i] = this._samples[index];
        }
        result.sort((left, right) => {
            return left - right;
        });
        this.onComputeComplete(this._length, result);
        return result;
    }
    /**
     * Whether the buffer has reached capacity
     */
    get isFull() {
        return this._length === this._capacity;
    }
    /**
     * Number of samples in the buffer
     *
     * Getter required: _length is mutated internally but exposed read-only
     */
    get length() {
        // Ensure non-negative (defensive - _length should never be negative)
        const result = Math.max(this._length, EMPTY_LENGTH);
        return result;
    }
    /**
     * Called at the start of clear(), before state is reset.
     * No-op default — override to observe buffer clears.
     */
    onClear() {
        // no-op
    }
    /**
     * Called in the full-buffer push path, immediately before the oldest
     * sample is overwritten. Not called when the buffer is not full.
     * No-op default — override to observe evictions.
     *
     * @param _oldValue - The sample value that will be overwritten
     */
    onEvict(_oldValue) {
        // no-op
    }
    /**
     * Called at the start of the full-buffer push path, before any eviction.
     * Receives the incoming sample value that triggered the overflow.
     * No-op default — override to observe buffer overflow events.
     *
     * @param _value - The incoming sample value that caused the overflow
     */
    onOverflow(_value) { }
    /**
     * Called at the end of push(), after the sample has been stored and
     * length/head updated.
     * No-op default — override to observe pushes.
     *
     * @param _value - The sample value that was pushed
     * @param _evicted - Whether an existing sample was evicted (buffer was full)
     */
    onPush(_value, _evicted) {
        // no-op
    }
    /**
     * Called immediately before returning from percentile(), after the
     * result has been calculated.
     * Not called when the buffer is empty (undefined return path).
     * No-op default — override to observe percentile calculations.
     *
     * @param _pct - The percentile requested (0-100)
     * @param _result - The calculated result value
     */
    onPercentile(_pct, _result) {
        // no-op
    }
    /**
     * Called at the start of buildSortedSamples(), before sorting.
     * Fires on a cache miss in percentile() — indicates a sort is about to run.
     * No-op default — override to observe sort/compute start events.
     *
     * @param _length - Number of samples about to be sorted
     */
    onComputeStart(_length) { }
    /**
     * Called at the end of buildSortedSamples(), after sorting is complete.
     * Fires on a cache miss in percentile() — the result is the freshly sorted array.
     * No-op default — override to observe sort/compute completion.
     *
     * @param _length - Number of samples that were sorted
     * @param _sorted - The sorted sample array (do not mutate)
     */
    onComputeComplete(_length, _sorted) { }
    /**
     * Calculate a percentile from the buffered samples using linear interpolation
     *
     * @param pct - Percentile to calculate (0-100)
     * @returns Percentile value or undefined if buffer is empty
     */
    percentile(pct) {
        if (this._length === EMPTY_LENGTH) {
            return undefined;
        }
        this._sortedCache ??= this.buildSortedSamples();
        const sorted = this._sortedCache;
        let result;
        // Handle edge cases
        if (pct <= EMPTY_LENGTH) {
            result = sorted[FIRST_ARRAY_INDEX];
            this.onPercentile(pct, result);
            return result;
        }
        if (pct >= PERCENTILE_MAX) {
            result = sorted.at(LAST_ARRAY_INDEX);
            this.onPercentile(pct, result);
            return result;
        }
        // Linear interpolation for percentile
        const rank = (pct / PERCENTILE_MAX) * (sorted.length - INCREMENT_BY_ONE);
        const lowerIndex = Math.floor(rank);
        const upperIndex = Math.ceil(rank);
        if (lowerIndex === upperIndex) {
            result = sorted[lowerIndex];
            this.onPercentile(pct, result);
            return result;
        }
        const fraction = rank - lowerIndex;
        const lowerValue = sorted[lowerIndex];
        const upperValue = sorted[upperIndex];
        result = lowerValue + fraction * (upperValue - lowerValue);
        this.onPercentile(pct, result);
        return result;
    }
    /**
     * Add a numeric sample to the buffer
     *
     * If the buffer is full, the oldest sample is overwritten.
     * Fire point: `onEvict` fires before overwrite; `onPush` fires at end.
     *
     * @param value - Sample value
     */
    push(value) {
        // Invalidate sorted cache
        this._sortedCache = null;
        if (this._length < this._capacity) {
            // Buffer not full - append
            this._samples[this._length] = value;
            this._length++;
            this.onPush(value, false);
        }
        else {
            // Buffer full - overwrite oldest (at head position)
            this.onOverflow(value);
            const oldValue = this._samples[this._head];
            this.onEvict(oldValue);
            this._samples[this._head] = value;
            this._head = (this._head + INCREMENT_BY_ONE) % this._capacity;
            this.onPush(value, true);
        }
    }
}
