// @studnicky/fsm/dist/EffectInterpreter.js
import { EffectInterpreterBuilder } from './EffectInterpreterBuilder.js';
import { FsmConfigError } from './errors/FsmConfigError.js';
import { InterpreterNotRunningError } from './errors/InterpreterNotRunningError.js';
import { InterpreterNotStartedError } from './errors/InterpreterNotStartedError.js';
export class EffectInterpreter {
    static create(options) {
        if (options.machine === undefined) {
            throw new FsmConfigError('machine is required');
        }
        if (options.machineId !== undefined && options.machineId === '') {
            throw new FsmConfigError('machineId must not be empty');
        }
        return new EffectInterpreter({
            'handlers': options.handlers,
            'machine': options.machine,
            'machineId': options.machineId
        });
    }
    static builder() {
        const result = EffectInterpreterBuilder.create((opts) => {
            const instance = EffectInterpreter.create(opts);
            return instance;
        });
        return result;
    }
    #machine;
    #handlers;
    #machineId;
    #observers = new Set();
    #mailbox = [];
    #currentState = undefined;
    #running = false;
    #draining = false;
    constructor(options) {
        this.#machine = options.machine;
        this.#handlers = options.handlers ?? ({});
        this.#machineId = options.machineId ?? crypto.randomUUID();
    }
    start() {
        if (this.#running) {
            return;
        }
        this.#currentState = this.#machine.getInitialState();
        this.#running = true;
        this.#notifyObservers();
        this.onStart(this.#currentState);
    }
    stop() {
        const state = this.#currentState;
        this.#running = false;
        this.onStop(state);
    }
    getState() {
        if (this.#currentState === undefined) {
            throw new InterpreterNotStartedError(`EffectInterpreter '${this.#machineId}' not started — call start() first`);
        }
        return this.#currentState;
    }
    async send(event) {
        if (!this.#running) {
            throw new InterpreterNotRunningError(`EffectInterpreter '${this.#machineId}' not running — call start() first`);
        }
        this.onEnqueue(event);
        this.#mailbox.push(event);
        if (!this.#draining) {
            await this.#drain();
        }
    }
    subscribe(observer) {
        this.#observers.add(observer);
        return () => { this.#observers.delete(observer); };
    }
    // ---------------------------------------------------------------------------
    // Lifecycle hooks — no-op by default. The bare class does NO observability;
    // override to add logging/tracing/metrics. Overrides must not throw or block.
    // ---------------------------------------------------------------------------
    /** Fires when the interpreter starts and the initial state is set. */
    onStart(_state) { }
    /** Fires when the interpreter is stopped. Receives the last known state. */
    onStop(_state) { }
    /** Fires when an event is enqueued in the mailbox. */
    onEnqueue(_event) { }
    /** Fires when the interpreter transitions to a new state variant. */
    onTransition(_from, _to, _event) { }
    /** Fires when entering a new state variant (after commit). */
    onEnterState(_state) { }
    /** Fires when exiting a state variant (before commit). */
    onExitState(_state) { }
    /** Fires before invoking an effect handler. */
    onEffectStart(_effect) { }
    /** Fires after an effect handler resolves successfully. */
    onEffectSuccess(_effect) { }
    /** Fires when an effect handler throws. */
    onEffectError(_effect, _error) { }
    async #drain() {
        this.#draining = true;
        while (this.#mailbox.length > 0) {
            const event = this.#mailbox.shift();
            if (event === undefined) {
                break;
            }
            const currentState = this.#currentState;
            if (currentState === undefined) {
                break;
            }
            const step = this.#machine.transition(currentState, event);
            const prevState = currentState;
            this.#currentState = step.state;
            if (prevState.variant !== step.state.variant) {
                this.onExitState(prevState);
                this.onTransition(prevState, step.state, event);
                this.onEnterState(step.state);
            }
            this.#notifyObservers();
            for (const effect of step.effects) {
                const variantKey = effect.variant;
                const handler = this.#handlers[variantKey];
                if (handler !== undefined) {
                    await this.#invokeHandler(effect, handler);
                }
            }
        }
        this.#draining = false;
    }
    async #invokeHandler(effect, handler) {
        this.onEffectStart(effect);
        try {
            await handler(effect);
            this.onEffectSuccess(effect);
        }
        catch (err) {
            const error = err instanceof Error ? err : new Error(String(err));
            this.onEffectError(effect, error);
            throw error;
        }
    }
    #notifyObservers() {
        const state = this.#currentState;
        if (state === undefined) {
            return;
        }
        for (const observer of this.#observers) {
            observer(state);
        }
    }
}
