// @studnicky/clock/dist/clock/Clock.js
import { ClockError } from '../errors/ClockError.js';
import { ClockBuilder } from './ClockBuilder.js';
const HRTIME_ZERO = 0n;
/**
 * Time source instance. Delegates to a `ClockProvider` (real or virtual)
 * while enforcing per-instance monotonicity for `now()` and `hrtime()`.
 */
export class Clock {
    static builder() {
        const result = ClockBuilder.create((provider) => {
            const clock = Clock.create(provider);
            return clock;
        });
        return result;
    }
    static create(provider) {
        return new this(provider);
    }
    #provider;
    #lastHrtime;
    #lastNow;
    /**
     * Property write order: #provider, #lastHrtime, #lastNow.
     */
    constructor(provider) {
        if (!Clock.isValidProvider(provider)) {
            throw new ClockError('provider must implement ClockProviderType');
        }
        this.#provider = provider;
        this.#lastHrtime = HRTIME_ZERO;
        this.#lastNow = 0;
    }
    static isValidProvider(provider) {
        return (typeof provider === 'object' &&
            provider !== null &&
            typeof provider.now === 'function' &&
            typeof provider.hrtime === 'function');
    }
    /**
     * Extension seam: subclasses may override to intercept or replace the raw
     * hrtime value before monotonicity clamping is applied.
     */
    readHrtime() {
        const result = this.#provider.hrtime();
        return result;
    }
    /**
     * Extension seam: subclasses may override to intercept or replace the raw
     * now value before monotonicity clamping is applied.
     */
    readNow() {
        const result = this.#provider.now();
        return result;
    }
    // ---------------------------------------------------------------------------
    // Lifecycle hooks — no-op by default. Override to add logging/tracing/metrics.
    // Overrides must not throw or block.
    // ---------------------------------------------------------------------------
    /**
     * Fires after each `now()` call, with the monotonically-clamped epoch-ms
     * value that was returned to the caller.
     */
    onNow(_timestamp) { }
    /**
     * Fires after each `hrtime()` call, with the monotonically-clamped
     * nanosecond bigint value that was returned to the caller.
     */
    onHrtime(_value) { }
    /**
     * Returns a monotonic nanosecond timestamp from the underlying provider.
     * Never returns a value smaller than the previous call on this instance.
     */
    hrtime() {
        const candidate = this.readHrtime();
        if (candidate > this.#lastHrtime) {
            this.#lastHrtime = candidate;
        }
        this.onHrtime(this.#lastHrtime);
        return this.#lastHrtime;
    }
    /**
     * Returns the current epoch-ms from the underlying provider.
     * Never returns a value smaller than the previous call on this instance.
     */
    now() {
        const candidate = this.readNow();
        if (candidate > this.#lastNow) {
            this.#lastNow = candidate;
        }
        this.onNow(this.#lastNow);
        return this.#lastNow;
    }
}
