// @studnicky/clock/dist/clock/VirtualClockProvider.js
import { ClockError } from '../errors/ClockError.js';
import { VirtualClockProviderBuilder } from './VirtualClockProviderBuilder.js';
/** Named constant: nanoseconds per millisecond, as BigInt. */
const NS_PER_MS = 1000000n;
/**
 * `ClockProvider` backed by a `VirtualTimeCounter`.
 * `now()` returns the counter's current epoch-ms.
 * `hrtime()` returns the same value converted to nanoseconds.
 */
export class VirtualClockProvider {
    static builder() {
        const result = VirtualClockProviderBuilder.create((counter) => {
            const provider = VirtualClockProvider.create(counter);
            return provider;
        });
        return result;
    }
    static create(counter) {
        return new this(counter);
    }
    #counter;
    /**
     * Property write order: #counter.
     */
    constructor(counter) {
        if (!VirtualClockProvider.isValidCounter(counter)) {
            throw new ClockError('counter must be a VirtualTimeCounter instance');
        }
        this.#counter = counter;
    }
    static isValidCounter(counter) {
        return (typeof counter === 'object' &&
            counter !== null &&
            typeof counter.nowMs === 'function' &&
            typeof counter.advance === 'function');
    }
    /**
     * Extension seam: exposes the injected counter to subclasses without widening
     * the public API surface.
     */
    get counter() {
        const result = this.#counter;
        return result;
    }
    /**
     * Extension seam: subclasses may override to replace or instrument the raw
     * virtual ms value before it is consumed by `hrtime()` and `now()`.
     */
    readVirtualMs() {
        const result = this.#counter.nowMs();
        return result;
    }
    // ---------------------------------------------------------------------------
    // Lifecycle hooks — no-op by default. Override to add logging/tracing/metrics.
    // Overrides must not throw or block.
    // ---------------------------------------------------------------------------
    /**
     * Fires after each `now()` call, with the virtual epoch-ms value (clamped
     * to 0 if the counter is negative) that was returned to the caller.
     */
    onNow(_timestamp) { }
    /**
     * Fires after each `hrtime()` call, with the virtual nanosecond bigint
     * value that was returned to the caller.
     */
    onHrtime(_value) { }
    /** Returns the virtual time in nanoseconds (epoch-ms * 1,000,000). */
    hrtime() {
        const result = BigInt(this.readVirtualMs()) * NS_PER_MS;
        this.onHrtime(result);
        return result;
    }
    /** Returns the virtual epoch-ms (always non-negative). */
    now() {
        const ms = this.readVirtualMs();
        const result = ms >= 0 ? ms : 0;
        this.onNow(result);
        return result;
    }
}
