// @studnicky/json/dist/json/Merge.js
/**
 * Merge — deep merge with V8-monomorphic write order.
 *
 * Semantics:
 * - Primitives: overlay wins when defined; base preserved on overlay = undefined.
 * - Arrays: overlay replaces base atomically (override `mergeArrays` to change).
 * - Objects: key-wise recursive merge; alphabetical union write order (monomorphic).
 * - Mismatched shapes (object vs array, primitive vs object): overlay wins entirely.
 * - Null is treated as a primitive (overlay-wins, not deep-merged).
 *
 * Subclass `Merge` and override protected static steps to customise merge behaviour.
 */
export class Merge {
    // ---------------------------------------------------------------------------
    // Protected steps — override in subclasses to customise merging
    // ---------------------------------------------------------------------------
    /** Return `true` when `value` is a mergeable plain object (not null, not array). */
    static isMergeable(value) {
        return typeof value === 'object' && value !== null && !Array.isArray(value);
    }
    /** Return the sorted union of keys from `base` and `overlay`. */
    static unionKeys(base, overlay) {
        const seenKeys = {};
        for (const baseKey of Object.keys(base)) {
            seenKeys[baseKey] = true;
        }
        for (const overlayKey of Object.keys(overlay)) {
            seenKeys[overlayKey] = true;
        }
        return Object.keys(seenKeys).sort();
    }
    /**
     * Merge two arrays.
     *
     * Default: overlay replaces base atomically.
     * Override to union, concat, or otherwise combine arrays.
     */
    static mergeArrays(_base, overlay) {
        const result = overlay;
        return result;
    }
    /** Merge two plain objects key-by-key in alphabetical union order. */
    static mergeObjects(base, overlay) {
        const keys = this.unionKeys(base, overlay);
        const merged = {};
        for (const key of keys) {
            merged[key] = this.mergeValues(base[key], overlay[key]);
        }
        return merged;
    }
    /**
     * Recursively merge two values. Dispatches to `mergeArrays` or
     * `mergeObjects` as appropriate; scalar overlay always wins.
     */
    static mergeValues(baseValue, overlayValue) {
        if (overlayValue === undefined) {
            return baseValue;
        }
        if (baseValue === undefined) {
            return overlayValue;
        }
        if (Array.isArray(overlayValue)) {
            if (Array.isArray(baseValue)) {
                return this.mergeArrays(baseValue, overlayValue);
            }
            return overlayValue;
        }
        if (Array.isArray(baseValue)) {
            return overlayValue;
        }
        if (!this.isMergeable(overlayValue)) {
            return overlayValue;
        }
        if (!this.isMergeable(baseValue)) {
            return overlayValue;
        }
        return this.mergeObjects(baseValue, overlayValue);
    }
    // ---------------------------------------------------------------------------
    // Private helpers
    // ---------------------------------------------------------------------------
    static assertDeepMergeResult(value) {
        void value;
    }
    // ---------------------------------------------------------------------------
    // Public static API
    // ---------------------------------------------------------------------------
    /**
     * Deeply merge `overlayValue` onto `baseValue` and return a new value.
     *
     * Overlay wins on conflicting primitives. Arrays replaced atomically.
     * Plain objects merged key-wise in alphabetical union order (monomorphic).
     */
    static deep(baseValue, overlayValue) {
        const result = this.mergeValues(baseValue, overlayValue);
        this.assertDeepMergeResult(result);
        return result;
    }
}
