// @studnicky/json/dist/json/Patch.js
/**
 * Patch — RFC-6902 JSON Patch operations.
 *
 * Supports: add, remove, replace, move, copy, test.
 * Paths must be JSON Pointer strings (RFC-6901): `/foo/bar`, `/items/0`.
 *
 * Subclass `Patch` and override any `protected` step to customise patch
 * behaviour. Static factories delegate through `this.make()` so a subclass
 * factory returns the subclass type.
 *
 * The default singleton `patch` is sufficient for most uses (prefer `Patch`
 * static factories for one-shot operations).
 */
import { PatchError } from '../errors/PatchError.js';
const VALID_OPS = new Set(['add', 'copy', 'move', 'remove', 'replace', 'test']);
export class Patch {
    operations;
    /**
     * Canonical entry point — validates operations and returns a `Patch` instance.
     *
     * Subclasses inherit this as `SubClass.create(...)` and receive the subclass
     * type because `new this(...)` resolves to the receiver's concrete class.
     */
    static create(operations = []) {
        return new this(operations);
    }
    constructor(operations = []) {
        const ops = Array.isArray(operations)
            ? operations
            : [operations];
        const opsLen = ops.length;
        for (let i = 0; i < opsLen; i += 1) {
            const operation = ops[i];
            if (!VALID_OPS.has(operation.op)) {
                throw new PatchError(`Invalid patch operation "${String(operation.op)}"; must be one of: ${[...VALID_OPS].join(', ')}`, String(operation.op), operation.path);
            }
        }
        this.operations = ops;
    }
    // ---------------------------------------------------------------------------
    // Protected factory seam — subclasses override to return their own type
    // ---------------------------------------------------------------------------
    /**
     * Create a new `Patch` instance from a list of operations.
     *
     * Subclasses override this to return an instance of the subclass, enabling
     * static factory methods (`add`, `remove`, etc.) to return the correct type
     * when called on the subclass constructor.
     */
    static make(ops) {
        const result = new this(ops);
        return result;
    }
    // ---------------------------------------------------------------------------
    // Static factory methods
    // ---------------------------------------------------------------------------
    /** Create a patch that adds `value` at `path`. */
    static add(path, value) {
        const result = this.make({ 'op': 'add', 'path': path, 'value': value });
        return result;
    }
    /** Combine multiple patches into a single patch. */
    static combine(...patches) {
        const ops = patches.flatMap((p) => {
            const patchOps = [...p.operations];
            return patchOps;
        });
        return this.make(ops);
    }
    /** Create a patch that copies a value from `from` to `path`. */
    static copy(from, path) {
        const result = this.make({ 'from': from, 'op': 'copy', 'path': path });
        return result;
    }
    /** Deserialize a patch from a plain-object representation. */
    static fromPlain(plain) {
        const result = this.make(plain.operations);
        return result;
    }
    /** Create a patch that moves a value from `from` to `path`. */
    static move(from, path) {
        const result = this.make({ 'from': from, 'op': 'move', 'path': path });
        return result;
    }
    /** Create a patch that removes the value at `path`. */
    static remove(path) {
        const result = this.make({ 'op': 'remove', 'path': path });
        return result;
    }
    /** Create a patch that replaces the value at `path` with `value`. */
    static replace(path, value) {
        const result = this.make({ 'op': 'replace', 'path': path, 'value': value });
        return result;
    }
    /** Create a patch that tests the value at `path` equals `value`. */
    static test(path, value) {
        const result = this.make({ 'op': 'test', 'path': path, 'value': value });
        return result;
    }
    // ---------------------------------------------------------------------------
    // Instance methods
    // ---------------------------------------------------------------------------
    /**
     * Apply this patch to `target` (mutates target in-place, per RFC-6902).
     *
     * Throws `PatchError` if any operation cannot be applied.
     */
    apply(target) {
        for (const op of this.operations) {
            this.applyOperation(target, op);
        }
        return target;
    }
    /** Return `true` when the patch has no operations. */
    isEmpty() {
        return this.operations.length === 0;
    }
    /** Return a copy of the operations array. */
    getOperations() {
        const ops = [...this.operations];
        return ops;
    }
    /** Serialize to a plain object. */
    toPlain() {
        const plain = { 'operations': this.operations };
        return plain;
    }
    /** Human-readable summary of operations. */
    toString() {
        const result = this.operations
            .map((op) => { const result = this.describeOp(op); return result; })
            .join(', ');
        return result;
    }
    // ---------------------------------------------------------------------------
    // Protected implementation steps — override to customise behaviour
    // ---------------------------------------------------------------------------
    /** Parse a JSON Pointer string into path segments. */
    parsePath(path) {
        if (!path.startsWith('/') && path !== '') {
            throw new PatchError(`Path must start with /: ${path}`, 'parsePath', path);
        }
        if (path === '' || path === '/') {
            return [];
        }
        return path.slice(1).split('/').map((part) => { const result = part.replace(/~1/gu, '/').replace(/~0/gu, '~'); return result; });
    }
    /** Read the value at `path` from `target`. */
    getValue(target, path) {
        const parts = this.parsePath(path);
        let current = target;
        const partsLen = parts.length;
        for (let i = 0; i < partsLen; i++) {
            const part = parts[i];
            if (current === null || current === undefined) {
                throw new PatchError(`Path not found: ${path}`, 'getValue', path);
            }
            if (typeof current !== 'object') {
                throw new PatchError(`Path not found: ${path}`, 'getValue', path);
            }
            const asRecord = current;
            if (!(part in asRecord)) {
                throw new PatchError(`Path not found: ${path}`, 'getValue', path);
            }
            current = asRecord[part];
        }
        return current;
    }
    /** Return `true` when `path` resolves to a value in `target`. */
    hasValue(target, path) {
        try {
            this.getValue(target, path);
            return true;
        }
        catch {
            return false;
        }
    }
    /** Write `value` at `path` in `target`, creating intermediate objects. */
    setValue(target, path, value) {
        const parts = this.parsePath(path);
        if (parts.length === 0) {
            // Replace root — not applicable to in-place mutation; caller must handle
            return;
        }
        let current = target;
        for (let i = 0; i < parts.length - 1; i++) {
            const part = parts[i];
            if (part === undefined) {
                break;
            }
            if (current === null || typeof current !== 'object') {
                throw new PatchError(`Intermediate path not traversable: ${path}`, 'setValue', path);
            }
            const asRecord = current;
            if (!(part in asRecord)) {
                asRecord[part] = {};
            }
            current = asRecord[part];
        }
        const lastPart = parts.at(-1);
        if (lastPart === undefined) {
            return;
        }
        if (current === null || typeof current !== 'object') {
            throw new PatchError(`Cannot set on non-object at: ${path}`, 'setValue', path);
        }
        current[lastPart] = value;
    }
    /** Remove the value at `path` from `target`. */
    removeValue(target, path) {
        const parts = this.parsePath(path);
        if (parts.length === 0) {
            return;
        }
        let current = target;
        for (let i = 0; i < parts.length - 1; i++) {
            const part = parts[i];
            if (part === undefined) {
                break;
            }
            if (current === null || typeof current !== 'object') {
                throw new PatchError(`Path not found: ${path}`, 'removeValue', path);
            }
            const asRecord = current;
            if (!(part in asRecord)) {
                throw new PatchError(`Path not found: ${path}`, 'removeValue', path);
            }
            current = asRecord[part];
        }
        const lastPart = parts.at(-1);
        if (lastPart === undefined) {
            return;
        }
        if (current === null || typeof current !== 'object') {
            throw new PatchError(`Cannot remove from non-object: ${path}`, 'removeValue', path);
        }
        if (Array.isArray(current)) {
            const idx = parseInt(lastPart, 10);
            current.splice(idx, 1);
        }
        else {
            const asRecord = current;
            Reflect.deleteProperty(asRecord, lastPart);
        }
    }
    /** Apply a single RFC-6902 `add` operation. */
    applyAdd(target, op) {
        this.setValue(target, op.path, op.value);
    }
    /** Apply a single RFC-6902 `replace` operation. */
    applyReplace(target, op) {
        if (!this.hasValue(target, op.path)) {
            throw new PatchError(`Cannot replace non-existent path: ${op.path}`, op.op, op.path);
        }
        this.setValue(target, op.path, op.value);
    }
    /** Apply a single RFC-6902 `remove` operation. */
    applyRemove(target, op) {
        this.removeValue(target, op.path);
    }
    /** Apply a single RFC-6902 `copy` operation. */
    applyCopy(target, op) {
        if (op.from === undefined) {
            throw new PatchError('copy operation requires "from"', op.op, op.path);
        }
        const copied = this.getValue(target, op.from);
        this.setValue(target, op.path, copied);
    }
    /** Apply a single RFC-6902 `move` operation. */
    applyMove(target, op) {
        if (op.from === undefined) {
            throw new PatchError('move operation requires "from"', op.op, op.path);
        }
        const moved = this.getValue(target, op.from);
        this.removeValue(target, op.from);
        this.setValue(target, op.path, moved);
    }
    /** Apply a single RFC-6902 `test` operation. */
    applyTest(target, op) {
        const actual = this.getValue(target, op.path);
        if (actual !== op.value) {
            throw new PatchError(`Test failed at ${op.path}: expected ${String(op.value)}, got ${String(actual)}`, op.op, op.path);
        }
    }
    static operationAppliers = {
        'add': (self, target, op) => { self.applyAdd(target, op); },
        'copy': (self, target, op) => { self.applyCopy(target, op); },
        'move': (self, target, op) => { self.applyMove(target, op); },
        'remove': (self, target, op) => { self.applyRemove(target, op); },
        'replace': (self, target, op) => { self.applyReplace(target, op); },
        'test': (self, target, op) => { self.applyTest(target, op); }
    };
    /** Apply a single RFC-6902 operation to `target`. */
    applyOperation(target, op) {
        const applier = Patch.operationAppliers[op.op];
        if (applier === undefined) {
            throw new PatchError(`Unknown patch operation: ${String(op.op)}`, String(op.op), op.path);
        }
        applier(this, target, op);
    }
    /** Produce a human-readable description of a single operation. */
    describeOp(op) {
        switch (op.op) {
            case 'add':
                return `ADD ${op.path} = ${JSON.stringify(op.value)}`;
            case 'copy':
                return `COPY ${op.from ?? '?'} → ${op.path}`;
            case 'move':
                return `MOVE ${op.from ?? '?'} → ${op.path}`;
            case 'remove':
                return `REMOVE ${op.path}`;
            case 'replace':
                return `REPLACE ${op.path} = ${JSON.stringify(op.value)}`;
            case 'test':
                return `TEST ${op.path} = ${JSON.stringify(op.value)}`;
            default:
                return `${String(op.op).toUpperCase()} ${op.path}`;
        }
    }
}
