// @studnicky/fetch/dist/modules/FetchClientBuilder.js
/**
 * Fluent builder for FetchClient
 */
/**
 * Fluent builder for constructing FetchClient instances
 */
export class FetchClientBuilder {
    static create(create) {
        return new FetchClientBuilder(create);
    }
    #create;
    #autoGenerateRequestId;
    #baseURL;
    #dispatcher;
    #headers;
    #metadata;
    #name;
    #params;
    #requestIdGenerator;
    #timeout;
    constructor(create) {
        this.#create = create;
    }
    withAutoGenerateRequestId(value) {
        this.#autoGenerateRequestId = value;
        return this;
    }
    withBaseURL(value) {
        this.#baseURL = value;
        return this;
    }
    withDispatcher(value) {
        this.#dispatcher = value;
        return this;
    }
    withHeaders(value) {
        this.#headers = value;
        return this;
    }
    withMetadata(value) {
        this.#metadata = value;
        return this;
    }
    withName(value) {
        this.#name = value;
        return this;
    }
    withParams(value) {
        this.#params = value;
        return this;
    }
    withRequestIdGenerator(value) {
        this.#requestIdGenerator = value;
        return this;
    }
    withTimeout(value) {
        this.#timeout = value;
        return this;
    }
    build() {
        const config = {};
        if (this.#autoGenerateRequestId !== undefined) {
            config.autoGenerateRequestId = this.#autoGenerateRequestId;
        }
        if (this.#baseURL !== undefined) {
            config.baseURL = this.#baseURL;
        }
        if (this.#dispatcher !== undefined) {
            config.dispatcher = this.#dispatcher;
        }
        if (this.#headers !== undefined) {
            config.headers = this.#headers;
        }
        if (this.#metadata !== undefined) {
            config.metadata = this.#metadata;
        }
        if (this.#name !== undefined) {
            config.name = this.#name;
        }
        if (this.#params !== undefined) {
            config.params = this.#params;
        }
        if (this.#requestIdGenerator !== undefined) {
            config.requestIdGenerator = this.#requestIdGenerator;
        }
        if (this.#timeout !== undefined) {
            config.timeout = this.#timeout;
        }
        return this.#create(config);
    }
}
