// @studnicky/fetch/dist/modules/RequestBuilder.js
/**
 * Fluent request builder for chaining request configuration
 */
import { UrlUtils } from './UrlUtils.js';
/**
 * Fluent request builder for chaining request configuration
 */
export class RequestBuilder {
    static create(client, path) {
        return new RequestBuilder(client, path);
    }
    client;
    fetchOptions = {};
    path;
    queryParams;
    requestBody;
    constructor(client, path) {
        this.client = client;
        this.path = path;
    }
    /**
     * Sets request body
     *
     * @param body - Request body
     * @returns This builder for chaining
     */
    body(body) {
        this.requestBody = body;
        return this;
    }
    /**
     * Builds final fetch options
     */
    buildOptions() {
        return { ...this.fetchOptions };
    }
    /**
     * Builds final path with query parameters
     */
    buildPath() {
        if (this.queryParams === undefined) {
            return this.path;
        }
        return UrlUtils.buildUrl(this.path, this.queryParams);
    }
    /**
     * Executes a DELETE request
     *
     * @returns Response promise
     */
    async delete() {
        const path = this.buildPath();
        const options = this.buildOptions();
        return await this.client.delete(path, options);
    }
    /**
     * Executes a GET request
     *
     * @returns Response promise
     */
    async get() {
        const path = this.buildPath();
        const options = this.buildOptions();
        return await this.client.get(path, options);
    }
    /**
     * Executes a HEAD request
     *
     * @returns Response promise
     */
    async head() {
        const path = this.buildPath();
        const options = this.buildOptions();
        return await this.client.head(path, options);
    }
    /**
     * Sets a single header
     *
     * @param name - Header name
     * @param value - Header value
     * @returns This builder for chaining
     */
    header(name, value) {
        const newHeaders = { ...this.fetchOptions.headers };
        newHeaders[name] = value;
        this.fetchOptions.headers = newHeaders;
        return this;
    }
    /**
     * Sets multiple headers
     *
     * @param headers - Headers to set
     * @returns This builder for chaining
     */
    headers(headers) {
        this.fetchOptions.headers = {
            ...this.fetchOptions.headers,
            ...headers
        };
        return this;
    }
    /**
     * Sets JSON body and Content-Type header
     *
     * @param data - Data to serialize as JSON
     * @returns This builder for chaining
     */
    json(data) {
        this.requestBody = data;
        this.fetchOptions.headers = {
            ...this.fetchOptions.headers,
            'Content-Type': 'application/json'
        };
        return this;
    }
    /**
     * Sets request metadata for logging and tracking
     *
     * Merged with client-level metadata
     * Accessible in lifecycle hooks via metadata.metadata
     *
     * @param metadata - Key-value pairs for tracking
     * @returns This builder for chaining
     *
     * @example
     * ```typescript
     * await client.request('/users')
     *   .metadata({ operation: 'fetchUsers', source: 'dashboard' })
     *   .get();
     * ```
     */
    metadata(metadata) {
        this.fetchOptions.metadata = {
            ...this.fetchOptions.metadata,
            ...metadata
        };
        return this;
    }
    /**
     * Executes an OPTIONS request
     *
     * @returns Response promise
     */
    async options() {
        const path = this.buildPath();
        const options = this.buildOptions();
        return await this.client.options(path, options);
    }
    /**
     * Builds options for body-bearing methods, merging fetchOptions with the stored request body
     */
    buildBodyOptions() {
        const opts = this.buildOptions();
        opts.body = this.requestBody;
        return opts;
    }
    /**
     * Executes a PATCH request
     *
     * @returns Response promise
     */
    async patch() {
        const result = this.client.patch(this.buildPath(), this.buildBodyOptions());
        return await result;
    }
    /**
     * Executes a POST request
     *
     * @returns Response promise
     */
    async post() {
        const result = this.client.post(this.buildPath(), this.buildBodyOptions());
        return await result;
    }
    /**
     * Executes a PUT request
     *
     * @returns Response promise
     */
    async put() {
        const result = this.client.put(this.buildPath(), this.buildBodyOptions());
        return await result;
    }
    /**
     * Adds a query string parameter
     *
     * Supports repeated keys by calling multiple times with the same key:
     * ```typescript
     * builder
     *   .queryString('tag', 'foo')
     *   .queryString('tag', 'bar')  // Results in: ?tag=foo&tag=bar
     * ```
     *
     * @param key - Query parameter key
     * @param value - Query parameter value (string, number, boolean, null, undefined, or array)
     * @returns This builder for chaining
     */
    queryString(key, value) {
        this.queryParams ??= {};
        const existingValue = this.queryParams[key];
        // If key already exists, convert to array or append to existing array
        if (existingValue === undefined) {
            // New key, set directly
            this.queryParams[key] = value;
        }
        else if (Array.isArray(existingValue)) {
            // Already an array, append new value(s)
            this.queryParams[key] = Array.isArray(value)
                ? [
                    ...existingValue,
                    ...value
                ]
                : [
                    ...existingValue,
                    value
                ];
        }
        else {
            // Convert to array with existing and new value(s)
            this.queryParams[key] = Array.isArray(value)
                ? [
                    existingValue,
                    ...value
                ]
                : [
                    existingValue,
                    value
                ];
        }
        return this;
    }
    /**
     * Sets custom request ID
     *
     * Overrides auto-generated request ID
     * Useful for passing request IDs from upstream services
     *
     * @param id - Request ID
     * @returns This builder for chaining
     *
     * @example
     * ```typescript
     * await client.request('/users')
     *   .requestId('req_from_upstream_123')
     *   .get();
     * ```
     */
    requestId(id) {
        this.fetchOptions.requestId = id;
        return this;
    }
    /**
     * Sets abort signal
     *
     * @param signal - AbortSignal
     * @returns This builder for chaining
     */
    signal(signal) {
        this.fetchOptions.signal = signal;
        return this;
    }
    /**
     * Sets request timeout
     *
     * @param ms - Timeout in milliseconds
     * @returns This builder for chaining
     */
    timeout(ms) {
        this.fetchOptions.timeout = ms;
        return this;
    }
}
