// @studnicky/fetch/dist/modules/UndiciDispatcherBuilder.js
/**
 * Fluent builder for UndiciDispatcher
 */
/**
 * Fluent builder for constructing UndiciDispatcher instances
 */
export class UndiciDispatcherBuilder {
    static create(create) {
        return new UndiciDispatcherBuilder(create);
    }
    #create;
    #allowH2;
    #bodyTimeout;
    #connections;
    #connectTimeout;
    #headersTimeout;
    #keepAliveTimeout;
    #localAddress;
    #maxConcurrentStreams;
    #maxResponseSize;
    #pipelining;
    constructor(create) {
        this.#create = create;
    }
    withAllowH2(value) {
        this.#allowH2 = value;
        return this;
    }
    withBodyTimeout(value) {
        this.#bodyTimeout = value;
        return this;
    }
    withConnections(value) {
        this.#connections = value;
        return this;
    }
    withConnectTimeout(value) {
        this.#connectTimeout = value;
        return this;
    }
    withHeadersTimeout(value) {
        this.#headersTimeout = value;
        return this;
    }
    withKeepAliveTimeout(value) {
        this.#keepAliveTimeout = value;
        return this;
    }
    withLocalAddress(value) {
        this.#localAddress = value;
        return this;
    }
    withMaxConcurrentStreams(value) {
        this.#maxConcurrentStreams = value;
        return this;
    }
    withMaxResponseSize(value) {
        this.#maxResponseSize = value;
        return this;
    }
    withPipelining(value) {
        this.#pipelining = value;
        return this;
    }
    build() {
        const config = {};
        if (this.#allowH2 !== undefined) {
            config.allowH2 = this.#allowH2;
        }
        if (this.#bodyTimeout !== undefined) {
            config.bodyTimeout = this.#bodyTimeout;
        }
        if (this.#connections !== undefined) {
            config.connections = this.#connections;
        }
        if (this.#connectTimeout !== undefined) {
            config.connectTimeout = this.#connectTimeout;
        }
        if (this.#headersTimeout !== undefined) {
            config.headersTimeout = this.#headersTimeout;
        }
        if (this.#keepAliveTimeout !== undefined) {
            config.keepAliveTimeout = this.#keepAliveTimeout;
        }
        if (this.#localAddress !== undefined) {
            config.localAddress = this.#localAddress;
        }
        if (this.#maxConcurrentStreams !== undefined) {
            config.maxConcurrentStreams = this.#maxConcurrentStreams;
        }
        if (this.#maxResponseSize !== undefined) {
            config.maxResponseSize = this.#maxResponseSize;
        }
        if (this.#pipelining !== undefined) {
            config.pipelining = this.#pipelining;
        }
        return this.#create(config);
    }
}
