// @studnicky/retry/dist/retry/core/ErrorClassifier.js
import { matchers } from './matchers.js';
/**
 * Abstract base class for error classification
 *
 * Provides a reusable pattern for implementing error classification logic
 * that can be used with retry mechanisms. Extend this class to create
 * protocol-specific or domain-specific error classifiers.
 *
 * @example HTTP Error Classifier
 * ```typescript
 * class HttpErrorClassifier extends ErrorClassifier {
 *   classify(error: Error, attemptNumber: number): ErrorClassificationType {
 *     if ('status' in error) {
 *       const status = (error as { status: number }).status;
 *
 *       if (status === 429) {
 *         return { retryable: true, reason: 'Rate limited' };
 *       }
 *
 *       if (status >= 500) {
 *         return { retryable: true, reason: 'Server error' };
 *       }
 *     }
 *
 *     return { retryable: false };
 *   }
 * }
 * ```
 *
 * @example SPARQL Error Classifier
 * ```typescript
 * class SparqlErrorClassifier extends ErrorClassifier {
 *   constructor(private readonly adapterName: string) {
 *     super();
 *   }
 *
 *   classify(error: Error, attemptNumber: number): ErrorClassificationType {
 *     const msg = error.message.toLowerCase();
 *
 *     if (msg.includes('transaction') || msg.includes('503')) {
 *       return { retryable: true, reason: `${this.adapterName} transient error` };
 *     }
 *
 *     if (msg.includes('syntax')) {
 *       return { retryable: false, reason: 'SPARQL syntax error' };
 *     }
 *
 *     return { retryable: false };
 *   }
 * }
 * ```
 */
export class ErrorClassifier {
    static ARRAY_MATCHERS = matchers.array;
    static BOOLEAN_MATCHERS = matchers.boolean;
    static DATABASE_MATCHERS = matchers.database;
    static HTTP_MATCHERS = matchers.http;
    static INSTANCE_MATCHERS = matchers.instance;
    static LOGIC_MATCHERS = matchers.logic;
    static NETWORK_MATCHERS = matchers.network;
    /**
     * Composable matcher utilities for flexible property checking
     *
     * These matchers can be used with hasProperty() for expressive error classification.
     *
     * @example Using matchers
     * ```typescript
     * class HttpErrorClassifier extends ErrorClassifier {
     *   classify(error: Error): ErrorClassificationType {
     *     const { HTTP_MATCHERS, NETWORK_MATCHERS, LOGIC_MATCHERS } = ErrorClassifier;
     *
     *     // Use domain-specific matchers
     *     if (this.hasProperty(error, 'status', HTTP_MATCHERS.isServerError)) {
     *       return this.retryable('Server error');
     *     }
     *
     *     // Compose with logical operators
     *     if (this.hasProperty(error, 'status', LOGIC_MATCHERS.or(HTTP_MATCHERS.isRateLimited, HTTP_MATCHERS.isGatewayError))) {
     *       return this.retryable('Retryable HTTP error');
     *     }
     *
     *     // Use network matchers
     *     if (this.hasProperty(error, 'code', NETWORK_MATCHERS.isConnectionError)) {
     *       return this.retryable('Connection error');
     *     }
     *
     *     return this.nonRetryable('Unknown error');
     *   }
     * }
     * ```
     */
    static NUMBER_MATCHERS = matchers.number;
    static OBJECT_MATCHERS = matchers.object;
    static PROTO_MATCHERS = matchers.proto;
    static STRING_MATCHERS = matchers.string;
    constructor() { }
    hasProperty(error, propertyName, matcher) {
        if (!(propertyName in error)) {
            return false;
        }
        const value = error[propertyName];
        if (matcher === undefined) {
            return true;
        }
        if (typeof matcher === 'function') {
            return matcher(value);
        }
        if (Array.isArray(matcher)) {
            return matcher.includes(value);
        }
        return value === matcher;
    }
    /**
     * Helper: Check if error message contains any of the specified strings (case-insensitive)
     *
     * @param error - The error to check
     * @param patterns - String patterns to search for
     * @returns True if error message contains any pattern
     *
     * @example
     * ```typescript
     * if (this.messageContains(error, 'timeout', 'connection refused')) {
     *   return this.retryable('Network error');
     * }
     * ```
     */
    messageContains(error, ...patterns) {
        const msg = error.message.toLowerCase();
        const lowerPatterns = patterns.map((pattern) => {
            const result = pattern.toLowerCase();
            return result;
        });
        return lowerPatterns.some((pattern) => {
            const result = msg.includes(pattern);
            return result;
        });
    }
    /**
     * Helper: Create a non-retryable classification
     *
     * @param reason - Reason for classification
     * @returns Non-retryable error classification
     */
    nonRetryable(reason) {
        const classification = {
            'reason': reason,
            'retryable': false
        };
        return classification;
    }
    /**
     * Helper: Create a retryable classification
     *
     * @param reason - Reason for classification
     * @returns Retryable error classification
     */
    retryable(reason) {
        const classification = {
            'reason': reason,
            'retryable': true
        };
        return classification;
    }
}
