// @studnicky/types/dist/guards/Empty.js
/**
 * Producers and emptiness predicates for the five core collection types.
 *
 * Static producer methods return fresh empty instances on every call so
 * callers never share a mutable reference. Static predicate methods are
 * strict: `isObject` returns `false` for `null`, arrays, Maps, and Sets;
 * `isArray` returns `false` for non-Array iterables; and so on.
 *
 * All methods are monomorphic and use consistent object shapes so V8 can
 * inline-cache them without deoptimisation.
 */
export class Empty {
    // ── Producers ────────────────────────────────────────────────────────────
    /** Returns a fresh empty string. */
    static string() {
        const result = '';
        return result;
    }
    /** Returns a fresh empty plain object typed as `Record<string, never>`. */
    static object() {
        const result = {};
        return result;
    }
    /** Returns a fresh empty array typed as `T[]`. */
    static array() {
        return [];
    }
    /** Returns a fresh empty `Map<K, V>`. */
    static map() {
        return new Map();
    }
    /** Returns a fresh empty `Set<T>`. */
    static set() {
        return new Set();
    }
    // ── Predicates ───────────────────────────────────────────────────────────
    /** Returns `true` when `value` is exactly the empty string `''`. */
    static isString(value) {
        return value === '';
    }
    /**
     * Returns `true` when `value` is a plain object with no own enumerable
     * keys. Returns `false` for `null`, arrays, Maps, Sets, and any other
     * non-plain-object value.
     */
    static isObject(value) {
        if (typeof value !== 'object' || value === null || Array.isArray(value)) {
            return false;
        }
        if (value instanceof Map || value instanceof Set) {
            return false;
        }
        return Object.keys(value).length === 0;
    }
    /** Returns `true` when `value` is an `Array` with `length === 0`. */
    static isArray(value) {
        return Array.isArray(value) && value.length === 0;
    }
    /** Returns `true` when `value` is a `Map` with `size === 0`. */
    static isMap(value) {
        return value instanceof Map && value.size === 0;
    }
    /** Returns `true` when `value` is a `Set` with `size === 0`. */
    static isSet(value) {
        return value instanceof Set && value.size === 0;
    }
}
