// @studnicky/file-lock/dist/FileLockBuilder.js
export class FileLockBuilder {
    static create(create) {
        return new FileLockBuilder(create);
    }
    #create;
    #fileSystem;
    #ownerToken;
    #path;
    #pollMs;
    #timeoutMs;
    constructor(create) {
        this.#create = create;
    }
    withFileSystem(value) {
        this.#fileSystem = value;
        return this;
    }
    withOwnerToken(value) {
        this.#ownerToken = value;
        return this;
    }
    withPath(value) {
        this.#path = value;
        return this;
    }
    withPollMs(value) {
        this.#pollMs = value;
        return this;
    }
    withTimeoutMs(value) {
        this.#timeoutMs = value;
        return this;
    }
    async build() {
        const options = {
            'path': this.#path ?? '',
            ...(this.#fileSystem !== undefined ? { 'fileSystem': this.#fileSystem } : {}),
            ...(this.#ownerToken !== undefined ? { 'ownerToken': this.#ownerToken } : {}),
            ...(this.#pollMs !== undefined ? { 'pollMs': this.#pollMs } : {}),
            ...(this.#timeoutMs !== undefined ? { 'timeoutMs': this.#timeoutMs } : {})
        };
        const result = await this.#create(options);
        return result;
    }
}
