// @studnicky/virtual-fs/dist/virtual-fs/VirtualFileSystem.js
import { VirtualFileSystemError } from '../errors/VirtualFileSystemError.js';
import { VirtualFileSystemBuilder } from './VirtualFileSystemBuilder.js';
const DEFAULT_CLOCK = {
    'hrtime': () => { return BigInt(Date.now()) * 1000000n; },
    'now': () => {
        const result = Date.now();
        return result;
    }
};
export class VirtualFileSystem {
    static builder() {
        const result = VirtualFileSystemBuilder.create((options) => {
            return new this(options);
        });
        return result;
    }
    static create(options) {
        const result = new this(options ?? {});
        return result;
    }
    #clock;
    #entries;
    #files;
    constructor(options) {
        this.#clock = options.clock ?? DEFAULT_CLOCK;
        this.#entries = new Map();
        this.#files = new Map();
        // Seed root directory
        this.#entries.set('/', { 'kind': 'directory', 'mtimeMs': this.#clock.now() });
        // Seed initial files if provided
        if (options.seed !== undefined) {
            const seeds = options.seed;
            const seedKeys = Array.from(seeds.keys());
            const seedLen = seedKeys.length;
            for (let i = 0; i < seedLen; i += 1) {
                const path = seedKeys[i];
                if (path !== undefined) {
                    const content = seeds.get(path);
                    if (content !== undefined) {
                        this.writeFileSync(path, content, 'utf8');
                    }
                }
            }
        }
    }
    onCreate(_path) { }
    onDelete(_path) { }
    onRead(_path) { }
    onRename(_oldPath, _newPath) { }
    onWrite(_path) { }
    existsSync(path) {
        return this.#files.has(path) || this.#entries.has(path);
    }
    mkdirSync(path, options) {
        const recursive = options?.recursive === true;
        const existingEntry = this.#entries.get(path);
        if (existingEntry?.kind === 'directory') {
            if (!recursive) {
                throw new VirtualFileSystemError(`EEXIST: directory already exists, mkdir '${path}'`);
            }
            return;
        }
        if (recursive) {
            const segments = path.split('/');
            const filtered = [];
            const segLen = segments.length;
            for (let i = 0; i < segLen; i += 1) {
                const s = segments[i];
                if (s !== undefined && s.length > 0) {
                    filtered.push(s);
                }
            }
            let current = '';
            const filteredLen = filtered.length;
            for (let i = 0; i < filteredLen; i += 1) {
                const seg = filtered[i];
                if (seg !== undefined) {
                    current = `${current}/${seg}`;
                    if (!this.#entries.has(current)) {
                        const entry = { 'kind': 'directory', 'mtimeMs': this.#clock.now() };
                        this.#entries.set(current, entry);
                        this.onCreate(current);
                    }
                }
            }
        }
        else {
            const entry = { 'kind': 'directory', 'mtimeMs': this.#clock.now() };
            this.#entries.set(path, entry);
            this.onCreate(path);
        }
    }
    readdirSync(path) {
        const entry = this.#entries.get(path);
        if (entry === undefined) {
            throw new VirtualFileSystemError(`ENOENT: no such file or directory, scandir '${path}'`);
        }
        if (entry.kind !== 'directory') {
            throw new VirtualFileSystemError(`ENOTDIR: not a directory, scandir '${path}'`);
        }
        const prefix = path === '/' ? '/' : `${path}/`;
        const result = [];
        const entryKeys = Array.from(this.#entries.keys());
        const entryLen = entryKeys.length;
        for (let i = 0; i < entryLen; i += 1) {
            const candidate = entryKeys[i];
            if (candidate !== undefined && candidate !== path && candidate.startsWith(prefix)) {
                const rest = candidate.slice(prefix.length);
                if (rest.length > 0 && !rest.includes('/')) {
                    result.push(rest);
                }
            }
        }
        const fileKeys = Array.from(this.#files.keys());
        const fileLen = fileKeys.length;
        for (let i = 0; i < fileLen; i += 1) {
            const candidate = fileKeys[i];
            if (candidate !== undefined &&
                !this.#entries.has(candidate) &&
                candidate.startsWith(prefix)) {
                const rest = candidate.slice(prefix.length);
                if (rest.length > 0 && !rest.includes('/')) {
                    result.push(rest);
                }
            }
        }
        this.onRead(path);
        return result;
    }
    readFileSync(path, _encoding) {
        const content = this.#files.get(path);
        if (content === undefined) {
            throw new VirtualFileSystemError(`ENOENT: no such file or directory, open '${path}'`);
        }
        this.onRead(path);
        return content;
    }
    renameSync(oldPath, newPath) {
        const content = this.#files.get(oldPath);
        if (content === undefined) {
            throw new VirtualFileSystemError(`ENOENT: no such file or directory, rename '${oldPath}' -> '${newPath}'`);
        }
        const oldEntry = this.#entries.get(oldPath);
        const mtimeMs = this.#clock.now();
        this.#files.set(newPath, content);
        this.#files.delete(oldPath);
        const kind = oldEntry !== undefined ? oldEntry.kind : 'file';
        this.#entries.set(newPath, { 'kind': kind, 'mtimeMs': mtimeMs });
        this.#entries.delete(oldPath);
        this.onRename(oldPath, newPath);
    }
    statSync(path) {
        const entry = this.#entries.get(path);
        const hasFile = this.#files.has(path);
        if (entry === undefined && !hasFile) {
            throw new VirtualFileSystemError(`ENOENT: no such file or directory, stat '${path}'`);
        }
        const kind = entry !== undefined ? entry.kind : 'file';
        const mtimeMs = entry !== undefined ? entry.mtimeMs : this.#clock.now();
        return {
            'isDirectory': () => { return kind === 'directory'; },
            'isFile': () => { return kind === 'file'; },
            'mtimeMs': mtimeMs
        };
    }
    unlinkSync(path) {
        if (!this.#files.has(path)) {
            throw new VirtualFileSystemError(`ENOENT: no such file or directory, unlink '${path}'`);
        }
        this.#files.delete(path);
        this.#entries.delete(path);
        this.onDelete(path);
    }
    writeFileSync(path, data, _encoding) {
        const isNew = !this.#files.has(path);
        const mtimeMs = this.#clock.now();
        this.#files.set(path, data);
        this.#entries.set(path, { 'kind': 'file', 'mtimeMs': mtimeMs });
        if (isNew) {
            this.onCreate(path);
        }
        else {
            this.onWrite(path);
        }
    }
}
