// @studnicky/cache/dist/LruCache.js
import { CacheConfigError } from './errors/index.js';
import { LruCacheBuilder } from './LruCacheBuilder.js';
const noExpiry = 0;
/**
 * In-process LRU + TTL cache with O(1) promotion on read.
 *
 * The base class performs NO observability of its own — it exposes protected
 * lifecycle hooks that a consumer overrides to add logging, timing, or metrics.
 * Overrides must not throw or block; hooks are called synchronously and
 * exceptions propagate to the caller.
 *
 * @example Adding observability via hooks
 * ```typescript
 * class ObservedCache extends LruCache<string, number> {
 *   protected override onHit(key: string, value: number): void {
 *     console.log(`[cache] hit key=${key} value=${value}`);
 *   }
 *   protected override onMiss(key: string): void {
 *     console.log(`[cache] miss key=${key}`);
 *   }
 * }
 * ```
 */
export class LruCache {
    #capacity;
    #defaultTtlMs;
    #entries;
    /** Maps internal (prefixed) string key → original caller key, for use in hooks. */
    #keyMap;
    #nodes;
    #prefix;
    #head;
    #tail;
    static create(options) {
        // `new this(...)` so subclass factories return the subclass instance.
        return new this(options);
    }
    static builder() {
        const factory = (options) => {
            const result = LruCache.create(options);
            return result;
        };
        const result = LruCacheBuilder.create(factory);
        return result;
    }
    constructor(options) {
        if (options.capacity <= 0 || !Number.isInteger(options.capacity)) {
            throw new CacheConfigError('capacity must be a positive integer');
        }
        this.#capacity = options.capacity;
        this.#defaultTtlMs = options.ttlMs;
        this.#entries = new Map();
        this.#keyMap = new Map();
        this.#nodes = new Map();
        this.#prefix = options.prefix ?? '';
        this.#head = undefined;
        this.#tail = undefined;
    }
    // ---------------------------------------------------------------------------
    // Lifecycle hooks — no-op by default. The bare class does NO observability;
    // override these to add logging, timing, or metrics. Overrides must not throw
    // or block; hooks fire synchronously after the relevant state mutation.
    // ---------------------------------------------------------------------------
    /** Fires when `get()` finds a live, non-expired entry. */
    onHit(_key, _value) { }
    /**
     * Fires when `get()` returns undefined — either because the key is absent or
     * because the entry was expired and lazily evicted. Use `onExpire` to
     * distinguish expired entries.
     */
    onMiss(_key) { }
    /** Fires when `set()` inserts a NEW key (key was not previously present). */
    onSet(_key) { }
    /** Fires when `set()` overwrites the value for an EXISTING key. */
    onUpdate(_key) { }
    /**
     * Fires when an entry is removed to make room for a new one (capacity
     * eviction of the LRU tail). The evicted key is the one removed, not the
     * one being inserted.
     */
    onEvict(_key, _reason) { }
    /**
     * Fires when `get()` or `has()` encounters an entry whose TTL has elapsed
     * and lazily removes it. Called before the entry is dropped, while the key
     * is still known.
     */
    onExpire(_key) { }
    /**
     * Fires when `delete()` removes an entry that actually existed. Not called
     * when delete() is a no-op for an absent key.
     */
    onDelete(_key) { }
    /**
     * Fires when `clear()` empties the cache. `count` is the number of entries
     * present (including expired-but-not-yet-evicted ones) before the wipe.
     */
    onClear(_count) { }
    /** Includes expired entries not yet lazily evicted. */
    get size() {
        const result = this.#entries.size;
        return result;
    }
    /** Promotes entry to MRU on hit; returns undefined on miss or expiry (lazily evicts). */
    get(key) {
        const internalKey = this.buildInternalKey(key);
        const entry = this.#entries.get(internalKey);
        if (entry === undefined) {
            this.onMiss(key);
            return undefined;
        }
        if (LruCache.isExpired(entry)) {
            this.onExpire(key);
            this.removeEntry(internalKey);
            this.onMiss(key);
            return undefined;
        }
        this.promoteToHead(internalKey);
        this.onHit(key, entry.value);
        return entry.value;
    }
    /**
     * Inserts each `[key, value]` pair in array order. The last entry in the array
     * ends up most-recently-used. An empty array is a no-op. `ttlMs` applies to all
     * entries in the batch with the same precedence as the per-call TTL in `set`.
     */
    setMany(entries, ttlMs) {
        for (const [key, value] of entries) {
            this.set(key, value, ttlMs);
        }
    }
    /** Stores value; promotes existing key to MRU or evicts LRU tail if at capacity. */
    set(key, value, ttlMs) {
        const internalKey = this.buildInternalKey(key);
        const effectiveTtl = ttlMs ?? this.#defaultTtlMs;
        const expiresAt = effectiveTtl !== undefined ? Date.now() + effectiveTtl : noExpiry;
        const existing = this.#entries.get(internalKey);
        if (existing !== undefined) {
            existing.expiresAt = expiresAt;
            existing.value = value;
            this.promoteToHead(internalKey);
            this.onUpdate(key);
            return;
        }
        if (this.#entries.size >= this.#capacity && this.#tail !== undefined) {
            this.evictTail(this.#tail.key);
        }
        const entry = {
            'expiresAt': expiresAt,
            'value': value
        };
        this.#entries.set(internalKey, entry);
        this.#keyMap.set(internalKey, key);
        const node = {
            'key': internalKey,
            'next': undefined,
            'prev': undefined
        };
        this.#nodes.set(internalKey, node);
        this.insertAtHead(node);
        this.onSet(key);
    }
    /** Returns true if key exists and has not expired; lazily evicts expired entries. */
    has(key) {
        const internalKey = this.buildInternalKey(key);
        const entry = this.#entries.get(internalKey);
        if (entry === undefined) {
            return false;
        }
        if (LruCache.isExpired(entry)) {
            this.onExpire(key);
            this.removeEntry(internalKey);
            return false;
        }
        return true;
    }
    delete(key) {
        const internalKey = this.buildInternalKey(key);
        const existed = this.#entries.has(internalKey);
        if (existed) {
            this.removeEntry(internalKey);
            this.onDelete(key);
        }
        return existed;
    }
    clear() {
        const count = this.#entries.size;
        this.#entries.clear();
        this.#keyMap.clear();
        this.#nodes.clear();
        this.#head = undefined;
        this.#tail = undefined;
        this.onClear(count);
    }
    static isExpired(entry) {
        if (entry.expiresAt === noExpiry) {
            return false;
        }
        return Date.now() > entry.expiresAt;
    }
    buildInternalKey(key) {
        const raw = String(key);
        return this.#prefix.length > 0 ? `${this.#prefix}:${raw}` : raw;
    }
    insertAtHead(node) {
        node.next = this.#head;
        node.prev = undefined;
        if (this.#head !== undefined) {
            this.#head.prev = node;
        }
        this.#head = node;
        if (this.#tail === undefined) {
            this.#tail = node;
        }
    }
    promoteToHead(key) {
        const node = this.#nodes.get(key);
        if (node === undefined || node === this.#head) {
            return;
        }
        this.unlinkNode(node);
        node.next = undefined;
        node.prev = undefined;
        this.insertAtHead(node);
    }
    unlinkNode(node) {
        if (node.prev !== undefined) {
            node.prev.next = node.next;
        }
        else {
            this.#head = node.next;
        }
        if (node.next !== undefined) {
            node.next.prev = node.prev;
        }
        else {
            this.#tail = node.prev;
        }
        node.next = undefined;
        node.prev = undefined;
    }
    removeEntry(key) {
        this.#entries.delete(key);
        this.#keyMap.delete(key);
        const node = this.#nodes.get(key);
        if (node !== undefined) {
            this.unlinkNode(node);
            this.#nodes.delete(key);
        }
    }
    evictTail(key) {
        const originalKey = this.#keyMap.get(key);
        this.#entries.delete(key);
        this.#keyMap.delete(key);
        const node = this.#nodes.get(key);
        if (node !== undefined) {
            this.unlinkNode(node);
            this.#nodes.delete(key);
        }
        if (originalKey !== undefined) {
            this.onEvict(originalKey, 'capacity');
        }
    }
}
