// @studnicky/fetch/dist/modules/browser/UndiciDispatcher.js
/**
 * Browser stub for UndiciDispatcher.
 * undici connection pooling requires a Node.js runtime.
 * Browser bundlers swap this file in place of the Node version via the
 * "browser" field in package.json.
 */
import { ConfigurationError } from '../../errors/index.js';
import { UndiciDispatcherBuilder } from '../UndiciDispatcherBuilder.js';
const BROWSER_ERROR_MESSAGE = 'undici connection pooling requires a Node.js runtime; the browser uses native fetch';
/**
 * Browser-environment stub. All instance methods throw ConfigurationError
 * because create() throws before any instance is constructed.
 */
export class UndiciDispatcher {
    static create(_config = {}) {
        throw new ConfigurationError(BROWSER_ERROR_MESSAGE);
    }
    static builder() {
        const result = UndiciDispatcherBuilder.create((_options) => {
            throw new ConfigurationError(BROWSER_ERROR_MESSAGE);
        });
        return result;
    }
    checkDispatcherHealth(_origin) {
        throw new ConfigurationError(BROWSER_ERROR_MESSAGE);
    }
    close() {
        throw new ConfigurationError(BROWSER_ERROR_MESSAGE);
    }
    destroy(_options) {
        throw new ConfigurationError(BROWSER_ERROR_MESSAGE);
    }
    getAgent() {
        throw new ConfigurationError(BROWSER_ERROR_MESSAGE);
    }
    getSignal() {
        throw new ConfigurationError(BROWSER_ERROR_MESSAGE);
    }
    getStats() {
        throw new ConfigurationError(BROWSER_ERROR_MESSAGE);
    }
}
