// @studnicky/errors/dist/errors/BaseError.js
import { CAUSE_CHAIN_DEPTH_LIMIT, CAUSE_DEPTH_SENTINEL } from '../constants/CauseChainConstants.js';
/**
 * Abstract base class for all errors in the system.
 * Subclasses must call `super(args)` and register their code via `ErrorCodeRegistry`.
 */
export class BaseError extends Error {
    /** Registered error code (dotted camelCase, e.g. `'errors.validationFailed'`). */
    code;
    /** Optional correlation ID for distributed tracing. */
    correlationId;
    /** Structured metadata dictionary attached to this error instance. */
    metadata;
    /** Whether this error represents a transient condition that may succeed on retry. */
    retryable;
    /** Unix millisecond timestamp at time of construction. */
    timestamp;
    constructor(args) {
        super(args.message, { 'cause': args.cause });
        // Property write order: name first (shadows Error.prototype.name).
        this.name = new.target.name;
        this.code = args.code;
        this.metadata = args.metadata !== undefined ? Object.freeze({ ...args.metadata }) : undefined;
        this.timestamp = Date.now();
        this.correlationId = args.correlationId;
        this.retryable = args.retryable ?? false;
    }
    /**
     * Finds the first cause of a specific `BaseError` subclass in the cause chain.
     * Returns `undefined` if not found.
     */
    static findCauseOfType(error, ctor) {
        let current = error;
        let depth = 0;
        while (current !== undefined && current !== null && depth < CAUSE_CHAIN_DEPTH_LIMIT) {
            if (current instanceof ctor) {
                return current;
            }
            if (current instanceof Error) {
                current = current.cause;
            }
            else {
                break;
            }
            depth++;
        }
        return undefined;
    }
    /**
     * Returns the full cause chain as a flat readonly array.
     * The first element is `error`; subsequent elements are causes.
     */
    static getCauseChain(error) {
        const chain = [];
        let current = error;
        let depth = 0;
        while (current !== undefined && current !== null && depth < CAUSE_CHAIN_DEPTH_LIMIT) {
            chain.push(current);
            if (current instanceof Error) {
                current = current.cause;
            }
            else {
                break;
            }
            depth++;
        }
        return chain;
    }
    /**
     * Returns `true` if any node in the cause chain is an instance of `ctor`.
     */
    static hasCauseOfType(error, ctor) {
        let current = error;
        let depth = 0;
        while (current !== undefined && current !== null && depth < CAUSE_CHAIN_DEPTH_LIMIT) {
            if (current instanceof ctor) {
                return true;
            }
            if (current instanceof Error) {
                current = current.cause;
            }
            else {
                break;
            }
            depth++;
        }
        return false;
    }
    /**
     * Returns the `message` of any error-like value as a string.
     * Used to safely interpolate caught `unknown` errors into messages.
     */
    static toMessage(err) {
        if (err instanceof Error) {
            return err.message;
        }
        return String(err);
    }
    /** Serializes a single cause node at the given depth. */
    static serializeCause(error, depth) {
        if (error instanceof BaseError) {
            const causeRaw = error.cause;
            let causeValue = null;
            if (causeRaw !== undefined && causeRaw !== null) {
                if (depth >= CAUSE_CHAIN_DEPTH_LIMIT) {
                    causeValue = CAUSE_DEPTH_SENTINEL;
                }
                else {
                    causeValue = BaseError.serializeCause(causeRaw, depth + 1);
                }
            }
            return {
                'cause': causeValue,
                'code': error.code,
                'context': error.metadata,
                'correlationId': error.correlationId ?? null,
                'message': error.message,
                'timestamp': error.timestamp
            };
        }
        if (error instanceof Error) {
            const causeRaw = error.cause;
            let causeValue = null;
            if (causeRaw !== undefined && causeRaw !== null) {
                if (depth >= CAUSE_CHAIN_DEPTH_LIMIT) {
                    causeValue = CAUSE_DEPTH_SENTINEL;
                }
                else {
                    causeValue = BaseError.serializeCause(causeRaw, depth + 1);
                }
            }
            return {
                'cause': causeValue,
                'code': 'native.error',
                'context': undefined,
                'correlationId': null,
                'message': error.message,
                'timestamp': 0
            };
        }
        // Primitive cause (string, number, etc.).
        return {
            'cause': null,
            'code': 'native.primitive',
            'context': undefined,
            'correlationId': null,
            'message': String(error),
            'timestamp': 0
        };
    }
    /**
     * Returns extra fields to merge into the `toJSON()` output.
     * Override in subclasses to inject additional serialized properties without
     * rewriting cause-chain logic.
     *
     * Fire-point: called from `toJSON()` after the base serialization object is built.
     * `super.serializeExtra()` need not be called — the default returns `{}`.
     */
    serializeExtra() {
        const extra = {};
        return extra;
    }
    /**
     * Formats the user-facing message string.
     * Override in subclasses to provide domain-specific formatting.
     * Fire-point: called from `toUserMessage()`.
     * `super.formatUserMessage()` need not be called — the default returns `this.message`.
     */
    formatUserMessage() {
        const message = this.message;
        return message;
    }
    /**
     * Serializes this error (and its cause chain) to a plain JSON-compatible object.
     * Circular cause chains are truncated at `CAUSE_CHAIN_DEPTH_LIMIT`.
     * Every field is always present; absent optional fields use `null` or `undefined`.
     *
     * Returns `Record<string, unknown>` so subclasses may override with richer
     * serialization while still satisfying the base contract.
     */
    toJSON() {
        const base = BaseError.serializeCause(this, 0);
        const extra = this.serializeExtra();
        return { ...base, ...extra };
    }
    /**
     * Serializes this error to a `SerializedErrorType` object.
     * Equivalent to `toJSON()` with the precise return type.
     */
    toSerializedError() {
        const result = BaseError.serializeCause(this, 0);
        return result;
    }
    /**
     * Returns a user-facing message string suitable for display.
     * Delegates to `formatUserMessage()` — override that method in subclasses.
     */
    toUserMessage() {
        const result = this.formatUserMessage();
        return result;
    }
}
