// @studnicky/system/dist/providers/browser/SystemProvider.js
import { GpuDetector } from '../../modules/browser/GpuDetector.js';
export class SystemProvider {
    arch() {
        const nav = globalThis.navigator;
        const raw = nav?.userAgentData?.platform ?? nav?.userAgent ?? '';
        const lower = raw.toLowerCase();
        if (lower.includes('mac') || lower.includes('iphone') || lower.includes('ipad')) {
            const result = 'arm64';
            return result;
        }
        const result = 'x64';
        return result;
    }
    cpuModel() {
        const result = 'Unknown';
        return result;
    }
    detectGpu() {
        const result = GpuDetector.detect();
        return result;
    }
    freeMb() {
        const result = 0;
        return result;
    }
    logicalCpuCount() {
        const nav = globalThis.navigator;
        const result = nav?.hardwareConcurrency ?? 1;
        return result;
    }
    platform() {
        const nav = globalThis.navigator;
        const raw = nav?.userAgentData?.platform ?? nav?.userAgent ?? '';
        const lower = raw.toLowerCase();
        if (lower.includes('mac') || lower.includes('iphone') || lower.includes('ipad')) {
            const result = 'darwin';
            return result;
        }
        if (lower.includes('win')) {
            const result = 'win32';
            return result;
        }
        if (lower.includes('linux') || lower.includes('android')) {
            const result = 'linux';
            return result;
        }
        const result = 'unknown';
        return result;
    }
    runtimeVersion() {
        const nav = globalThis.navigator;
        const result = nav?.userAgent ?? 'unknown';
        return result;
    }
    totalMb() {
        const nav = globalThis.navigator;
        const gb = nav?.deviceMemory;
        if (gb === undefined) {
            const result = 0;
            return result;
        }
        const result = Math.floor(gb * 1024);
        return result;
    }
}
