// @studnicky/timing/dist/modules/NoOpTiming.js
import { NoOpTimingBuilder } from './NoOpTimingBuilder.js';
/**
 * No-operation timing tracker that discards all events.
 *
 * Useful for:
 * - Testing when you don't want timing overhead
 * - Production environments where timing is disabled
 * - Libraries that want a default silent timer
 *
 * @public
 *
 * @example
 * ```typescript
 * import { NoOpTiming, TimingEvent } from '@studnicky/timing';
 *
 * const timing = NoOpTiming.create();
 *
 * // All timing calls are ignored
 * timing.event(TimingEvent.create()
 *   .component('GraphAdapter')
 *   .operation('query')
 *   .build());
 *
 * timing.getEvents(); // Returns { durationMs: 0 }
 * ```
 */
export class NoOpTiming {
    /**
     * Creates a new NoOpTiming instance.
     * Use `NoOpTiming.builder().build()` for a consistent builder API.
     *
     * @returns A new NoOpTiming instance
     *
     * @example
     * ```typescript
     * import { NoOpTiming, TimingEvent } from '@studnicky/timing';
     *
     * const timing = NoOpTiming.create();
     *
     * timing.event(TimingEvent.create()
     *   .component('GraphAdapter')
     *   .operation('query')
     *   .build()); // Does nothing
     * ```
     */
    static create() {
        return new this();
    }
    /**
     * Creates a new NoOpTimingBuilder for a uniform builder API.
     *
     * @returns A new NoOpTimingBuilder instance
     *
     * @example
     * ```typescript
     * import { NoOpTiming } from '@studnicky/timing';
     *
     * const timing = NoOpTiming.builder().build();
     * ```
     */
    static builder() {
        const result = NoOpTimingBuilder.create(() => {
            const instance = NoOpTiming.create();
            return instance;
        });
        return result;
    }
    /**
     * Protected constructor. Use NoOpTiming.create() to instantiate.
     */
    constructor() { }
    /**
     * No-operation clear method.
     * @returns this for method chaining
     */
    clear() {
        return this;
    }
    /**
     * No-operation event method.
     *
     * @param _data - Ignored event data
     */
    event(_data) {
        // No-op
    }
    /**
     * No-operation getEvents method.
     *
     * @returns Empty events with zero duration
     */
    getEvents() {
        const events = { 'durationMs': 0 };
        return events;
    }
}
