// @studnicky/logger/dist/modules/BaseLogEntryBuilder.js
import { LogBuildError } from '../errors/LogBuildError.js';
/**
 * Base class for log entry builders providing shared fluent API methods.
 */
export class BaseLogEntryBuilder {
    constructor() { }
    componentName;
    contextCalled = false;
    contextData = {};
    durationMs;
    operationName;
    statusValue;
    /**
     * Set the component name (e.g., 'graph', 'cache', 'api').
     * Required for build().
     */
    component(name) {
        this.componentName = name;
        this.onFieldSet('component', name);
        return this;
    }
    /**
     * Add contextual data.
     * Required for build() - use empty object if no context needed.
     * Context data is nested under a 'context' key in the output.
     */
    context(data) {
        this.contextCalled = true;
        Object.assign(this.contextData, data);
        this.onFieldSet('context', data);
        return this;
    }
    /**
     * Set the operation duration.
     */
    duration(ms) {
        this.durationMs = ms;
        this.onFieldSet('duration', ms);
        return this;
    }
    /**
     * Set the operation name (e.g., 'query', 'get', 'response').
     * Required for build().
     */
    operation(name) {
        this.operationName = name;
        this.onFieldSet('operation', name);
        return this;
    }
    /**
     * Set the operation status.
     * Required for build().
     */
    status(status) {
        this.statusValue = status;
        this.onFieldSet('status', status);
        return this;
    }
    /**
     * Hook called after each field is set via the fluent API.
     * Override in subclasses to observe field mutations.
     * Default implementation is a no-op.
     */
    onFieldSet(_field, _value) { }
    /**
     * Hook called after `build()` assembles the frozen result.
     * Override in subclasses to observe completed build results.
     * Default implementation is a no-op.
     */
    onBuild(_result) { }
    /**
     * Hook called when a required field is missing and `build()` is about to throw.
     * Override in subclasses to observe build errors before they are thrown.
     * Default implementation is a no-op.
     */
    onBuildError(_field) { }
    /**
     * Validates the fields shared by all log entry builders.
     * Throws LogBuildError if component, operation, status, or context are missing.
     * Subclasses call this and then validate their own additional required fields.
     */
    validateRequiredFields() {
        if (this.componentName === undefined) {
            this.onBuildError('component');
            throw new LogBuildError(`${this.constructor.name}: component is required`);
        }
        if (this.operationName === undefined) {
            this.onBuildError('operation');
            throw new LogBuildError(`${this.constructor.name}: operation is required`);
        }
        if (this.statusValue === undefined) {
            this.onBuildError('status');
            throw new LogBuildError(`${this.constructor.name}: status is required`);
        }
        if (!this.contextCalled) {
            this.onBuildError('context');
            throw new LogBuildError(`${this.constructor.name}: context is required (use empty object {} if no context needed)`);
        }
    }
}
