// @studnicky/logger/dist/modules/Logger.js
import { LogLevel } from '../constants/LogLevel.js';
import { configValidation } from './configValidation.js';
import { LoggerBuilder } from './LoggerBuilder.js';
import { parseLogLevel } from './parseLogLevel.js';
/**
 * Core logger with pluggable transport architecture.
 *
 * The global level acts as a floor — records below it are discarded before
 * reaching any transport. Each transport may apply its own, higher floor via
 * its `level` option. A Logger with no transports is valid and silent.
 *
 * @example
 * ```typescript
 * import { Logger, ConsoleTransport, MemoryTransport } from '@studnicky/logger';
 *
 * const memory = MemoryTransport.create();
 * const logger = Logger.create({
 *   level: 'debug',
 *   metadata: { service: 'api' },
 *   transports: [ConsoleTransport.create({ level: 'warn' }), memory]
 * });
 *
 * logger.info(body);   // reaches MemoryTransport only (ConsoleTransport floor is warn)
 * logger.warn(body);   // reaches both transports
 *
 * const child = logger.child({ requestId: 'abc' });
 * child.error(fault);  // metadata merged: { service: 'api', requestId: 'abc' }
 * ```
 */
export class Logger {
    /**
     * Creates a new Logger instance.
     *
     * @param options - Configuration for the logger
     * @returns A new Logger instance
     */
    static create(options = {}) {
        return new this(options);
    }
    static builder() {
        const result = LoggerBuilder.create((options) => { const result = Logger.create(options); return result; });
        return result;
    }
    #level;
    #metadata;
    #transports;
    constructor(options = {}) {
        configValidation.assertPlainObject(options.metadata, 'metadata');
        if (options.transports !== undefined && !Array.isArray(options.transports)) {
            throw new Error('transports must be an array');
        }
        this.#level = options.level !== undefined
            ? parseLogLevel(options.level)
            : LogLevel.INFO;
        this.#metadata = options.metadata ?? {};
        this.#transports = options.transports ?? [];
    }
    /**
     * Creates a child logger sharing the same level and transports with
     * metadata merged from the parent.
     *
     * @param metadata - Additional metadata to include in all child log records
     * @returns A new Logger with merged metadata
     */
    child(metadata) {
        const result = Logger.create({
            'level': this.#level,
            'metadata': { ...this.#metadata, ...metadata },
            'transports': this.#transports
        });
        this.onChildCreate(metadata);
        return result;
    }
    /**
     * Emits a debug record to all transports that meet the level floor.
     *
     * @param data - Structured log data from LogBody.build()
     */
    debug(data) {
        this.emit(LogLevel.DEBUG, data);
    }
    /**
     * Emits an error record to all transports that meet the level floor.
     *
     * @param data - Structured log data from LogBody.build() or LogFault.build()
     */
    error(data) {
        this.emit(LogLevel.ERROR, data);
    }
    /**
     * Emits an info record to all transports that meet the level floor.
     *
     * @param data - Structured log data from LogBody.build()
     */
    info(data) {
        this.emit(LogLevel.INFO, data);
    }
    /**
     * Emits a trace record to all transports that meet the level floor.
     *
     * @param data - Structured log data from LogBody.build()
     */
    trace(data) {
        this.emit(LogLevel.TRACE, data);
    }
    /**
     * Emits a warn record to all transports that meet the level floor.
     *
     * @param data - Structured log data from LogBody.build() or LogFault.build()
     */
    warn(data) {
        this.emit(LogLevel.WARN, data);
    }
    emit(level, data) {
        if (level < this.#level) {
            this.onDropped(level);
            return;
        }
        const record = {
            'data': data,
            'level': level,
            'metadata': this.#metadata,
            'time': Date.now()
        };
        this.onLog(level, record);
        const count = this.#transports.length;
        for (let i = 0; i < count; i += 1) {
            const transport = this.#transports[i];
            if (transport === undefined) {
                continue;
            }
            this.writeToTransport(transport, record);
        }
    }
    writeToTransport(transport, record) {
        try {
            transport.write(record);
        }
        catch (error) {
            // One failing transport must not prevent others from receiving the record.
            this.onTransportError(transport, record, error);
        }
    }
    /**
     * Hook called after a record is assembled and before fan-out to transports.
     * Override in subclasses to observe every emitted record.
     * Default implementation is a no-op.
     */
    onLog(_level, _record) { }
    /**
     * Hook called when a record is below the logger's level floor and is discarded.
     * Override in subclasses to observe dropped records.
     * Default implementation is a no-op.
     */
    onDropped(_level) { }
    /**
     * Hook called after a child logger is created via `.child()`.
     * Override in subclasses to observe child logger creation.
     * Default implementation is a no-op.
     */
    onChildCreate(_bindings) { }
    /**
     * Hook called when a transport's `write()` throws.
     * Override in subclasses to observe transport errors.
     * Default implementation is a no-op.
     */
    onTransportError(_transport, _record, _error) { }
}
