// @studnicky/event-bus/dist/BusQueueBuilder.js
/** Fluent builder for creating BusQueue instances. */
import { BusQueueConfigError } from './errors/index.js';
export class BusQueueBuilder {
    static create(create) {
        const result = new BusQueueBuilder(create);
        return result;
    }
    #create;
    #handler;
    #highWaterMark;
    #onError;
    #signal;
    constructor(create) {
        this.#create = create;
    }
    withHandler(handler) {
        this.#handler = handler;
        return this;
    }
    withHighWaterMark(value) {
        this.#highWaterMark = value;
        return this;
    }
    withOnError(callback) {
        this.#onError = callback;
        return this;
    }
    withSignal(signal) {
        this.#signal = signal;
        return this;
    }
    build() {
        const handler = this.#handler;
        if (handler === undefined) {
            throw new BusQueueConfigError('BusQueueBuilder: handler is required — call withHandler() before build()');
        }
        const options = {
            'handler': handler,
            ...(this.#highWaterMark !== undefined ? { 'highWaterMark': this.#highWaterMark } : {}),
            ...(this.#onError !== undefined ? { 'onError': this.#onError } : {}),
            ...(this.#signal !== undefined ? { 'signal': this.#signal } : {})
        };
        return this.#create(options);
    }
}
