// @studnicky/event-bus/dist/BusQueue.js
/** Bounded async FIFO queue with backpressure; enqueue blocks at highWaterMark. */
import { BusQueueBuilder } from './BusQueueBuilder.js';
import { BusQueueConfigError } from './errors/BusQueueConfigError.js';
const busQueueDefaultHighWaterMark = 256;
export class BusQueue {
    #handler;
    #hwm;
    #onError;
    #onEnqueueCb;
    #onDequeueCb;
    #onDropCb;
    #onOverflowCb;
    #onSlowConsumerCb;
    #onHandlerErrorCb;
    #queue = [];
    #backpressureWaiters = [];
    #drainWaiters = [];
    #draining = false;
    #aborted = false;
    static builder() {
        const result = BusQueueBuilder.create((options) => {
            const instance = BusQueue.create(options);
            return instance;
        });
        return result;
    }
    static create(options) {
        const result = new this(options);
        return result;
    }
    constructor(options) {
        if (typeof options.handler !== 'function') {
            throw new BusQueueConfigError('handler must be a function');
        }
        const hwmOption = options.highWaterMark;
        if (hwmOption !== undefined && (!Number.isInteger(hwmOption) || hwmOption <= 0)) {
            throw new BusQueueConfigError('highWaterMark must be a positive integer');
        }
        this.#handler = options.handler;
        this.#hwm = hwmOption ?? busQueueDefaultHighWaterMark;
        this.#onError = options.onError;
        this.#onEnqueueCb = options.onEnqueue;
        this.#onDequeueCb = options.onDequeue;
        this.#onDropCb = options.onDrop;
        this.#onOverflowCb = options.onOverflow;
        this.#onSlowConsumerCb = options.onSlowConsumer;
        this.#onHandlerErrorCb = options.onHandlerError;
        const signal = options.signal;
        if (signal !== undefined) {
            if (signal.aborted) {
                this.#aborted = true;
            }
            else {
                signal.addEventListener('abort', () => { this.#handleAbort(); }, { 'once': true });
            }
        }
    }
    get size() {
        const result = this.#queue.length;
        return result;
    }
    async enqueue(item) {
        if (this.#aborted) {
            this.onDrop();
            if (this.#onDropCb !== undefined) {
                this.#onDropCb();
            }
            return;
        }
        this.#queue.push(item);
        this.onEnqueue(this.#queue.length);
        if (this.#onEnqueueCb !== undefined) {
            this.#onEnqueueCb(this.#queue.length);
        }
        this.#scheduleLoop();
        if (this.#queue.length < this.#hwm) {
            return;
        }
        this.onOverflow(this.#queue.length);
        if (this.#onOverflowCb !== undefined) {
            this.#onOverflowCb(this.#queue.length);
        }
        this.onSlowConsumer(this.#queue.length, this.#hwm);
        if (this.#onSlowConsumerCb !== undefined) {
            this.#onSlowConsumerCb(this.#queue.length, this.#hwm);
        }
        await new Promise((resolve) => {
            this.#backpressureWaiters.push({ 'resolve': resolve });
        });
    }
    async drain() {
        if (this.#queue.length === 0 || this.#aborted) {
            return;
        }
        await new Promise((resolve) => {
            this.#drainWaiters.push({ 'resolve': resolve });
        });
    }
    #handleAbort() {
        this.#aborted = true;
        for (const w of this.#backpressureWaiters) {
            w.resolve();
        }
        this.#backpressureWaiters.length = 0;
        for (const w of this.#drainWaiters) {
            w.resolve();
        }
        this.#drainWaiters.length = 0;
    }
    #scheduleLoop() {
        if (this.#draining) {
            return;
        }
        queueMicrotask(() => { void this.#runDrainLoop(); });
    }
    async #tryHandleItem(item) {
        try {
            await this.#handler(item);
        }
        catch (err) {
            if (this.#onError !== undefined) {
                this.#onError(err);
            }
            this.onHandlerError(err);
            if (this.#onHandlerErrorCb !== undefined) {
                this.#onHandlerErrorCb(err);
            }
        }
    }
    async #runDrainLoop() {
        this.#draining = true;
        while (this.#queue.length > 0 && !this.#aborted) {
            const item = this.#queue.shift();
            if (item === undefined) {
                break;
            }
            const waiter = this.#backpressureWaiters.shift();
            if (waiter !== undefined) {
                waiter.resolve();
            }
            this.onDequeue(this.#queue.length);
            if (this.#onDequeueCb !== undefined) {
                this.#onDequeueCb(this.#queue.length);
            }
            await this.#tryHandleItem(item);
        }
        this.#draining = false;
        if (this.#queue.length === 0 || this.#aborted) {
            for (const w of this.#drainWaiters) {
                w.resolve();
            }
            this.#drainWaiters.length = 0;
        }
    }
    /** Fires when an item is added to the queue (after push). */
    onEnqueue(_depth) { }
    /** Fires when an item is removed from the queue for processing (after shift). */
    onDequeue(_depth) { }
    /** Fires when enqueue is called but the queue is already aborted (item silently dropped). */
    onDrop() { }
    /** Fires when queue depth reaches highWaterMark and caller will block (backpressure begins). */
    onOverflow(_depth) { }
    /** Fires when queue depth reaches highWaterMark — same moment as onOverflow; args include depth and hwm. */
    onSlowConsumer(_depth, _highWaterMark) { }
    /** Fires after handler throws and onError callback (if any) has been called. */
    onHandlerError(_error) { }
}
