// @studnicky/event-bus/dist/EventBus.js
/** Typed multi-topic pub/sub; per-subscriber BusQueue isolates errors and backpressure. */
import { BusQueue } from './BusQueue.js';
import { EventBusBuilder } from './EventBusBuilder.js';
export class EventBus {
    #store = new Map();
    #busController = new AbortController();
    static builder() {
        const result = EventBusBuilder.create(() => {
            const instance = EventBus.create();
            return instance;
        });
        return result;
    }
    static create() {
        const result = new this();
        return result;
    }
    constructor() { }
    #getTopicMap(topic) {
        const key = String(topic);
        let topicMap = this.#store.get(key);
        if (topicMap === undefined) {
            topicMap = new Map();
            this.#store.set(key, topicMap);
        }
        return topicMap;
    }
    subscribe(topic, handler, options) {
        const topicMap = this.#getTopicMap(topic);
        const queueController = new AbortController();
        const stopQueue = () => { queueController.abort(); };
        if (this.#busController.signal.aborted) {
            queueController.abort();
        }
        else {
            this.#busController.signal.addEventListener('abort', stopQueue, { 'once': true });
        }
        const callerSignal = options?.signal;
        if (callerSignal !== undefined) {
            if (callerSignal.aborted) {
                queueController.abort();
            }
            else {
                callerSignal.addEventListener('abort', stopQueue, { 'once': true });
            }
        }
        const queue = BusQueue.create({
            'handler': async (payload) => {
                await handler(payload, queueController.signal);
                this.onDeliver(topic, payload);
            },
            'onDequeue': (_depth) => { this.onDequeue(topic); },
            'onDrop': () => { this.onDrop(topic); },
            'onEnqueue': (_depth) => { this.onEnqueue(topic); },
            'onHandlerError': (err) => { this.onHandlerError(topic, err); },
            'onOverflow': (depth) => { this.onOverflow(topic, depth); },
            'onSlowConsumer': (depth, _hwm) => { this.onSlowConsumer(topic, depth); },
            'signal': queueController.signal
        });
        topicMap.set(handler, queue);
        this.onSubscribe(topic);
        let unsubscribed = false;
        return () => {
            if (unsubscribed) {
                return;
            }
            unsubscribed = true;
            this.#store.get(String(topic))?.delete(handler);
            queueController.abort();
            this.#busController.signal.removeEventListener('abort', stopQueue);
            this.onUnsubscribe(topic);
        };
    }
    async publish(topic, payload) {
        const topicMap = this.#store.get(String(topic));
        if (topicMap === undefined || topicMap.size === 0) {
            return;
        }
        this.onPublish(topic, payload);
        await Promise.all([...topicMap.values()].map((q) => { const result = q.enqueue(payload); return result; }));
    }
    async drain() {
        const promises = [];
        for (const topicMap of this.#store.values()) {
            for (const queue of topicMap.values()) {
                promises.push(queue.drain());
            }
        }
        await Promise.all(promises);
    }
    async close() {
        this.onDispose();
        this.#busController.abort();
        await this.drain();
    }
    /** Fires when publish() is called for a topic (once per publish, before fan-out). */
    onPublish(_topic, _payload) { }
    /** Fires when a subscriber is registered for a topic. */
    onSubscribe(_topic) { }
    /** Fires when a subscriber is removed (unsubscribe fn called or signal aborted). */
    onUnsubscribe(_topic) { }
    /** Fires after each individual event delivery to a handler (per-queue, per-event). */
    onDeliver(_topic, _payload) { }
    /** Fires when a subscriber queue is at or above highWaterMark (backpressure). */
    onSlowConsumer(_topic, _depth) { }
    /** Fires when a subscriber handler throws an error. */
    onHandlerError(_topic, _error) { }
    /** Fires when the bus is closed (bus.close() called). */
    onDispose() { }
    /** Fires when an event is enqueued to a subscriber queue. */
    onEnqueue(_topic) { }
    /** Fires when an event is dequeued from a subscriber queue for delivery. */
    onDequeue(_topic) { }
    /** Fires when an event is dropped (queue aborted / subscriber gone). */
    onDrop(_topic) { }
    /** Fires when overflow/backpressure begins on a subscriber queue. */
    onOverflow(_topic, _depth) { }
}
