// @studnicky/retry/dist/retry/backoff/BackoffStrategy.js
import { Guard } from '@studnicky/config';
import { EXPONENTIAL_BASE, JITTER_BASE, LINEAR_INCREMENT } from '../../constants/index.js';
const maxMultiplier = 32;
/** Backoff strategy computations as static methods. */
export class BackoffStrategy {
    static constant(_attemptNumber, baseDelayMs) {
        const result = baseDelayMs;
        return result;
    }
    static exponential(attemptNumber, baseDelayMs) {
        return baseDelayMs * Math.pow(EXPONENTIAL_BASE, attemptNumber);
    }
    static exponentialWithJitter(attemptNumber, baseDelayMs) {
        const exponentialDelay = baseDelayMs * Math.pow(EXPONENTIAL_BASE, attemptNumber);
        const jitter = JITTER_BASE + Math.random();
        return Math.floor(exponentialDelay * jitter);
    }
    static linear(attemptNumber, baseDelayMs) {
        return baseDelayMs * (attemptNumber + LINEAR_INCREMENT);
    }
    static decorrelatedJitter(attemptNumber, baseDelayMs) {
        const maxDelay = baseDelayMs * maxMultiplier;
        if (attemptNumber <= 0) {
            return baseDelayMs;
        }
        const prevDelay = Math.min(maxDelay, baseDelayMs * (3 ** attemptNumber));
        const ceiling = Math.min(maxDelay, prevDelay * 3);
        return baseDelayMs + Math.random() * (ceiling - baseDelayMs);
    }
    /** Wraps a strategy capping its output at ceilingMs. */
    static withCeiling(strategy, ceilingMs) {
        return (attempt, base) => { const result = Math.min(ceilingMs, strategy(attempt, base)); return result; };
    }
    static isValid(value) {
        const result = Guard.isFunction(value);
        return result;
    }
}
