// @studnicky/config/dist/validation/configValidation.js
/**
 * Configuration validation utilities.
 *
 * Static assertion methods for validating configuration options.
 * All methods skip validation if value is undefined or null.
 * All methods throw (via `onValidationError`) on validation failure.
 *
 * Subclass and `static override onValidationError` to change the thrown error type.
 */
import { Guard } from '@studnicky/types';
import { ConfigurationError } from '../errors/ConfigurationError.js';
export class ConfigValidation {
    /**
     * Extension seam — called by every assert method on validation failure.
     * Subclasses may `static override` to throw a domain-specific error instead.
     *
     * Fire-point: called with the constructed message string immediately before
     * the throw. `super.onValidationError` need not be called; the default throws
     * `ConfigurationError.create(message)`.
     */
    static onValidationError(message) {
        throw ConfigurationError.create(message);
    }
    static assertString(val, name) {
        if (val === undefined || val === null) {
            return;
        }
        if (!Guard.isString(val)) {
            this.onValidationError(`${name} must be a string`);
        }
    }
    static assertNumber(val, name) {
        if (val === undefined || val === null) {
            return;
        }
        if (!Guard.isNumber(val)) {
            this.onValidationError(`${name} must be a number`);
        }
    }
    static assertBoolean(val, name) {
        if (val === undefined || val === null) {
            return;
        }
        if (!Guard.isBoolean(val)) {
            this.onValidationError(`${name} must be a boolean`);
        }
    }
    static assertFunction(val, name) {
        if (val === undefined || val === null) {
            return;
        }
        if (!Guard.isFunction(val)) {
            this.onValidationError(`${name} must be a function`);
        }
    }
    /**
     * Assert number is an integer.
     * Assumes assertNumber has already passed.
     */
    static assertInteger(val, name) {
        if (val === undefined || val === null) {
            return;
        }
        if (!Number.isInteger(val)) {
            this.onValidationError(`${name} must be an integer`);
        }
    }
    /**
     * Assert number is finite (not Infinity or -Infinity).
     * Assumes assertNumber has already passed.
     */
    static assertFinite(val, name) {
        if (val === undefined || val === null) {
            return;
        }
        if (!Number.isFinite(val)) {
            this.onValidationError(`${name} must be finite`);
        }
    }
    /**
     * Assert number is non-negative (>= 0).
     * Assumes assertNumber has already passed.
     */
    static assertNonNegative(val, name) {
        if (val === undefined || val === null || typeof val !== 'number') {
            return;
        }
        if (val < 0) {
            this.onValidationError(`${name} must be non-negative`);
        }
    }
    /**
     * Assert number is positive (> 0).
     * Assumes assertNumber has already passed.
     */
    static assertPositive(val, name) {
        if (val === undefined || val === null || typeof val !== 'number') {
            return;
        }
        if (val <= 0) {
            this.onValidationError(`${name} must be positive`);
        }
    }
    /**
     * Assert number is at least min value.
     * Assumes assertNumber has already passed.
     */
    static assertMin(val, min, name) {
        if (val === undefined || val === null || typeof val !== 'number') {
            return;
        }
        if (val < min) {
            this.onValidationError(`${name} must be at least ${min}`);
        }
    }
    /**
     * Assert number is positive or Infinity.
     * Assumes assertNumber has already passed.
     */
    static assertPositiveOrInfinity(val, name) {
        if (val === undefined || val === null || typeof val !== 'number') {
            return;
        }
        if (val !== Infinity && val <= 0) {
            this.onValidationError(`${name} must be positive or Infinity`);
        }
    }
    static assertHasMethod(val, method, name) {
        if (val === undefined || val === null) {
            return;
        }
        if (!Guard.isObject(val)) {
            this.onValidationError(`${name} must be an object`);
            return;
        }
        try {
            if (!(method in val) || typeof (val)[method] !== 'function') {
                this.onValidationError(`${name} must have a ${method} method`);
            }
        }
        catch (error) {
            // Re-throw errors produced by onValidationError (already the intended error type).
            // Only errors from exotic property access (non-Error throws are possible) fall through.
            if (error instanceof Error) {
                throw error;
            }
            this.onValidationError(`${name} must have a ${method} method`);
        }
    }
    static assertFunctionOrObjectWithMethod(val, method, name) {
        if (val === undefined || val === null) {
            return;
        }
        if (Guard.isFunction(val)) {
            return;
        }
        if (Guard.isObject(val)) {
            try {
                if (method in val && typeof (val)[method] === 'function') {
                    return;
                }
            }
            catch {
                // Intentionally ignore errors during property access on exotic objects
            }
        }
        this.onValidationError(`${name} must be a function or an object with a ${method} method`);
    }
    static assertNoUnknownKeys(config, knownKeys) {
        for (const key of Object.keys(config)) {
            if (!knownKeys.has(key)) {
                this.onValidationError(`Unknown configuration key: ${key}`);
            }
        }
    }
}
