// nuxt-swiper/dist/module.mjs
import { defineNuxtModule, createResolver, addPlugin, addImportsDir } from '@nuxt/kit';

const name = "nuxt-swiper";
const version = "2.0.2";

const module$1 = defineNuxtModule({
  meta: {
    name,
    version,
    configKey: "swiper"
  },
  defaults: {
    enableComposables: true,
    bundled: true
  },
  setup(opts, nuxt) {
    const resolver = createResolver(import.meta.url);
    nuxt.options.appConfig = {
      ...nuxt.options.appConfig,
      __swiper: {
        bundled: opts?.bundled ?? true
      }
    };
    const isCustomElement = nuxt.options.vue.compilerOptions.isCustomElement;
    nuxt.options.vue.compilerOptions.isCustomElement = (tag) => tag.startsWith("swiper-") || isCustomElement?.(tag) || false;
    nuxt.options.vite ||= {};
    nuxt.options.vite.optimizeDeps ||= {};
    nuxt.options.vite.optimizeDeps.include ||= [];
    nuxt.options.vite.optimizeDeps.include.push(
      opts.bundled ? "swiper/element/bundle" : "swiper/element"
    );
    if (nuxt.options.vue.runtimeCompiler)
      addPlugin(resolver.resolve("./runtime/plugins/custom-elements"));
    if (opts.enableComposables) addImportsDir(resolver.resolve("./runtime/composables"));
    addPlugin(resolver.resolve("./runtime/plugins/components.client"));
    nuxt.hook("prepare:types", ({ references }) => {
      references.push({
        path: resolver.resolve("./runtime/components.d.ts")
      });
    });
  }
});

export { module$1 as default };
