// @studnicky/predicates/dist/Predicates.js
/** Static predicate library for JSON Schema draft 2020-12 validation. */
import { DataType } from '@studnicky/json';
/** Scaling factor applied to `Number.EPSILON` when testing `multipleOf`. */
const multipleOfEpsilonFactor = 10;
/** Content encodings actively validated at runtime. */
const supportedContentEncodings = new Set(['base64', 'base64url']);
/** Content media types actively validated at runtime. */
const supportedContentMediaTypes = new Set(['application/json']);
export class Predicates {
    static coercionHandlers = new Map([
        [
            'array',
            (value) => {
                return !Array.isArray(value) && typeof value !== 'object' ? [value] : value;
            }
        ],
        [
            'boolean',
            (value) => {
                if (typeof value === 'string') {
                    return Predicates.coerceToBoolean(value) ?? value;
                }
                if (value === 1) {
                    return true;
                }
                if (value === 0) {
                    return false;
                }
                return value;
            }
        ],
        [
            'integer',
            (value) => {
                if (typeof value === 'string') {
                    const coerced = Predicates.coerceToNumber(value);
                    return coerced === undefined ? value : Math.trunc(coerced);
                }
                if (typeof value === 'boolean') {
                    return value ? 1 : 0;
                }
                return value;
            }
        ],
        [
            'null',
            (value) => {
                return value === '' || value === 'null' ? null : value;
            }
        ],
        [
            'number',
            (value) => {
                if (typeof value === 'string') {
                    return Predicates.coerceToNumber(value) ?? value;
                }
                if (typeof value === 'boolean') {
                    return value ? 1 : 0;
                }
                return value;
            }
        ],
        [
            'string',
            (value) => {
                return typeof value === 'string' ? value : String(value);
            }
        ]
    ]);
    static typeMatchers = new Map([
        [
            'array',
            (value) => {
                const result = Array.isArray(value);
                return result;
            }
        ],
        [
            'integer',
            (value) => {
                const result = Predicates.isIntegerValue(value);
                return result;
            }
        ],
        [
            'null',
            (value) => {
                return value === null;
            }
        ],
        [
            'number',
            (value) => {
                const result = Predicates.isFiniteNumber(value);
                return result;
            }
        ],
        [
            'object',
            (value) => {
                return Predicates.inferValueType(value) === 'object';
            }
        ]
    ]);
    /** Count Unicode code points without allocating an intermediate array. */
    static codePointLength(str) {
        let length = 0;
        const strLen = str.length;
        for (let index = 0; index < strLen; index++) {
            length++;
            const code = str.codePointAt(index);
            if (code !== undefined && code > 0xFF_FF) {
                index++;
            }
        }
        return length;
    }
    /** Returns true as soon as `target` code points have been counted; stops early. */
    static codePointLengthAtLeast(str, target) {
        let count = 0;
        const strLen = str.length;
        for (let index = 0; index < strLen; index++) {
            count++;
            if (count >= target) {
                return true;
            }
            const code = str.codePointAt(index);
            if (code !== undefined && code > 0xFF_FF) {
                index++;
            }
        }
        return count >= target;
    }
    /** Returns false as soon as code-point count exceeds `limit`; stops early. */
    static codePointLengthAtMost(str, limit) {
        let count = 0;
        const strLen = str.length;
        for (let index = 0; index < strLen; index++) {
            count++;
            if (count > limit) {
                return false;
            }
            const code = str.codePointAt(index);
            if (code !== undefined && code > 0xFF_FF) {
                index++;
            }
        }
        return true;
    }
    /** Coerce string to boolean; returns true/false for recognised literals, undefined otherwise. */
    static coerceToBoolean(value) {
        if (value === 'true' || value === '1') {
            return true;
        }
        if (value === 'false' || value === '0') {
            return false;
        }
        return undefined;
    }
    /** Coerce string to finite number; returns undefined for Infinity, NaN, or non-numeric. */
    static coerceToNumber(value) {
        const coerced = Number(value);
        return Number.isFinite(coerced) ? coerced : undefined;
    }
    /** Attempt coercion in schema-type order; returns first successful result or original value. */
    static coerceValue(schemaTypes, value) {
        if (value === undefined || value === null || schemaTypes.length === 0) {
            return value;
        }
        const typesLen = schemaTypes.length;
        for (let i = 0; i < typesLen; i += 1) {
            const type = schemaTypes[i];
            const coercer = Predicates.coercionHandlers.get(type);
            if (coercer !== undefined) {
                const result = coercer(value);
                if (result !== value) {
                    return result;
                }
            }
        }
        return value;
    }
    /** Infer the JSON Schema type name of a value. */
    static inferValueType(value) {
        if (value === null) {
            return 'null';
        }
        if (Array.isArray(value)) {
            return 'array';
        }
        return typeof value;
    }
    static isFiniteNumber(value) {
        return typeof value === 'number' && Number.isFinite(value);
    }
    static isIntegerValue(value) {
        return typeof value === 'number' && Number.isInteger(value);
    }
    static matchesAnyType(schemaTypes, value) {
        const result = schemaTypes.some((schemaType) => {
            const result = Predicates.matchesType(schemaType, value);
            return result;
        });
        return result;
    }
    static matchesType(schemaType, value) {
        const matcher = Predicates.typeMatchers.get(schemaType);
        return matcher === undefined ? Predicates.inferValueType(value) === schemaType : matcher(value);
    }
    static satisfiesConst(value, constValue) {
        const result = DataType.deepEqual(value, constValue);
        return result;
    }
    static satisfiesEnum(value, enumValues) {
        const result = enumValues.some((enumValue) => {
            const result = DataType.deepEqual(value, enumValue);
            return result;
        });
        return result;
    }
    static checkMinimum(value, minimum, exclusive) {
        return exclusive ? value > minimum : value >= minimum;
    }
    static checkMaximum(value, maximum, exclusive) {
        return exclusive ? value < maximum : value <= maximum;
    }
    /** Uses epsilon tolerance for floating-point rounding errors. */
    static checkMultipleOf(value, divisor) {
        if (divisor === 0) {
            return false;
        }
        const quotient = value / divisor;
        return Math.abs(quotient - Math.round(quotient)) <= Number.EPSILON * multipleOfEpsilonFactor;
    }
    static satisfiesMinimum(value, minimum) {
        return value >= minimum;
    }
    static satisfiesMaximum(value, maximum) {
        return value <= maximum;
    }
    static satisfiesExclusiveMinimum(value, limit) {
        return value > limit;
    }
    static satisfiesExclusiveMaximum(value, limit) {
        return value < limit;
    }
    static satisfiesMultipleOf(value, divisor) {
        const result = Predicates.checkMultipleOf(value, divisor);
        return result;
    }
    /** Fast-paths: len<min→false, len>=2*min→true; walks code points only in residual band. */
    static checkMinLength(value, minimum) {
        const result = Predicates.satisfiesMinLength(value, minimum);
        return result;
    }
    /** Fast-path: code_points <= utf16_length, so value.length<=max is definitely true. */
    static checkMaxLength(value, maximum) {
        const result = Predicates.satisfiesMaxLength(value, maximum);
        return result;
    }
    static checkPattern(value, pattern) {
        const regex = pattern instanceof RegExp ? pattern : new RegExp(pattern, 'u');
        return regex.test(value);
    }
    static satisfiesMinLength(value, minimum) {
        const len = value.length;
        if (len < minimum) {
            return false;
        }
        if (len >= minimum * 2) {
            return true;
        }
        return Predicates.codePointLengthAtLeast(value, minimum);
    }
    static satisfiesMaxLength(value, maximum) {
        if (value.length <= maximum) {
            return true;
        }
        return Predicates.codePointLengthAtMost(value, maximum);
    }
    static satisfiesPattern(value, regex) {
        const result = regex.test(value);
        return result;
    }
    /** Only base64/base64url are actively checked; unknown encodings return true per spec. */
    static satisfiesContentEncoding(value, encoding) {
        if (!supportedContentEncodings.has(encoding)) {
            return true;
        }
        return Predicates.#decodeBase64Safe(value, encoding === 'base64url') !== null;
    }
    /** Only application/json is actively checked; unknown media types return true per spec. */
    static satisfiesContentMediaType(value, mediaType, encoding) {
        if (!supportedContentMediaTypes.has(mediaType)) {
            return true;
        }
        let content = value;
        if (encoding !== undefined && supportedContentEncodings.has(encoding)) {
            const decoded = Predicates.#decodeBase64Safe(value, encoding === 'base64url');
            if (decoded === null) {
                return false;
            }
            content = decoded;
        }
        if (mediaType === 'application/json') {
            return Predicates.#isValidJson(content);
        }
        return true;
    }
    static checkMinItems(value, minimum) {
        return value.length >= minimum;
    }
    static checkMaxItems(value, maximum) {
        return value.length <= maximum;
    }
    static checkUniqueItems(value) {
        const result = Predicates.satisfiesUniqueItems(value);
        return result;
    }
    /** Validates minContains/maxContains bounds against match count from a contains schema. */
    static satisfiesContains(matchCount, minContains, maxContains) {
        const minimum = minContains ?? (maxContains === undefined ? 1 : 0);
        if (matchCount < minimum) {
            return false;
        }
        if (maxContains !== undefined && matchCount > maxContains) {
            return false;
        }
        return true;
    }
    static satisfiesMinItems(value, minimum) {
        return value.length >= minimum;
    }
    static satisfiesMaxItems(value, maximum) {
        return value.length <= maximum;
    }
    static satisfiesUniqueItems(value) {
        const valueLen = value.length;
        for (let index = 0; index < valueLen; index++) {
            for (let other = index + 1; other < valueLen; other++) {
                if (DataType.deepEqual(value[index], value[other])) {
                    return false;
                }
            }
        }
        return true;
    }
    static checkRequired(value, required) {
        const result = Predicates.hasAllRequiredProperties(value, required);
        return result;
    }
    static hasAllRequiredProperties(value, required) {
        const result = required.every((key) => {
            return key in value;
        });
        return result;
    }
    static hasNoAdditionalProperties(value, allowedKeys) {
        const result = Object.keys(value).every((key) => {
            const result = allowedKeys.has(key);
            return result;
        });
        return result;
    }
    static satisfiesMinProperties(value, minimum) {
        return Object.keys(value).length >= minimum;
    }
    static satisfiesMaxProperties(value, maximum) {
        return Object.keys(value).length <= maximum;
    }
    static #decodeBase64Safe(value, urlSafe) {
        try {
            const normalised = urlSafe
                ? value.replaceAll('-', '+').replaceAll('_', '/')
                : value;
            const decoded = atob(normalised);
            return decoded;
        }
        catch {
            return null;
        }
    }
    static #isValidJson(content) {
        try {
            JSON.parse(content);
            return true;
        }
        catch {
            return false;
        }
    }
}
