// @studnicky/resilience/dist/CircuitBreakerBuilder.js
/** Fluent builder for CircuitBreaker. */
export class CircuitBreakerBuilder {
    #create;
    #failureThreshold;
    #resetTimeoutMs;
    #successThreshold;
    #name;
    #clock;
    static create(create) {
        return new CircuitBreakerBuilder(create);
    }
    constructor(create) {
        this.#create = create;
    }
    withFailureThreshold(value) {
        this.#failureThreshold = value;
        return this;
    }
    withResetTimeoutMs(value) {
        this.#resetTimeoutMs = value;
        return this;
    }
    withSuccessThreshold(value) {
        this.#successThreshold = value;
        return this;
    }
    withName(value) {
        this.#name = value;
        return this;
    }
    withClock(value) {
        this.#clock = value;
        return this;
    }
    build() {
        const options = {
            'failureThreshold': this.#failureThreshold ?? 1,
            'resetTimeoutMs': this.#resetTimeoutMs ?? 0,
            ...(this.#successThreshold !== undefined ? { 'successThreshold': this.#successThreshold } : {}),
            ...(this.#name !== undefined ? { 'name': this.#name } : {}),
            ...(this.#clock !== undefined ? { 'clock': this.#clock } : {})
        };
        return this.#create(options);
    }
}
