// @studnicky/resilience/dist/CircuitBreaker.js
/** Async circuit breaker: closed → open (on failure threshold) → halfOpen (on timeout) → closed. */
import { CircuitBreakerBuilder } from './CircuitBreakerBuilder.js';
import { CircuitBreakerOpenError } from './CircuitBreakerOpenError.js';
import { ResilienceConfigError } from './errors/ResilienceConfigError.js';
export class CircuitBreaker {
    #failureThreshold;
    #resetTimeoutMs;
    #successThreshold;
    #name;
    #clock;
    #state = 'closed';
    #failureCount = 0;
    #successCount = 0;
    #openedAt = 0;
    static builder() {
        const factory = (options) => {
            const result = CircuitBreaker.create(options);
            return result;
        };
        const result = CircuitBreakerBuilder.create(factory);
        return result;
    }
    static create(options) {
        return new this(options);
    }
    constructor(options) {
        if (options.failureThreshold < 1) {
            throw new ResilienceConfigError('failureThreshold must be >= 1');
        }
        if (options.resetTimeoutMs < 0) {
            throw new ResilienceConfigError('resetTimeoutMs must be >= 0');
        }
        this.#failureThreshold = options.failureThreshold;
        this.#resetTimeoutMs = options.resetTimeoutMs;
        this.#successThreshold = options.successThreshold ?? 1;
        this.#name = options.name ?? 'circuit-breaker';
        this.#clock = options.clock ?? (() => { const result = Date.now(); return result; });
    }
    get state() {
        const result = this.#state;
        return result;
    }
    async execute(fn) {
        this.#checkHalfOpen();
        if (this.#state === 'open') {
            this.onReject();
            throw new CircuitBreakerOpenError(this.#name);
        }
        try {
            const result = await fn();
            this.#recordSuccess();
            return result;
        }
        catch (err) {
            this.#recordFailure(err);
            throw err;
        }
    }
    reset() {
        this.#state = 'closed';
        this.#failureCount = 0;
        this.#successCount = 0;
        this.#openedAt = 0;
        this.onClose();
    }
    forceClosed() { this.reset(); }
    forceOpen() {
        this.#state = 'open';
        this.#openedAt = this.#clock();
        this.onOpen();
    }
    /**
     * Fires after `fn()` resolves successfully in any circuit state.
     * Override to add logging, metrics, or tracing. Must not throw or block.
     */
    onSuccess() { }
    /**
     * Fires after `fn()` throws in any circuit state.
     * Override to add logging, metrics, or tracing. Must not throw or block.
     */
    onFailure(_error) { }
    /**
     * Fires when the failure threshold is reached and the circuit transitions closed → open.
     * Does NOT fire on the halfOpen → open re-open path. Must not throw or block.
     */
    onTrip() { }
    /**
     * Fires every time the circuit state becomes open (threshold trip or halfOpen → open on failure).
     * Must not throw or block.
     */
    onOpen() { }
    /**
     * Fires when the circuit transitions open → halfOpen after resetTimeoutMs.
     * Must not throw or block.
     */
    onHalfOpen() { }
    /**
     * Fires when the circuit becomes closed (successThreshold reached in halfOpen or manual reset).
     * Must not throw or block.
     */
    onClose() { }
    /**
     * Fires when execute() short-circuits because the circuit is open.
     * Must not throw or block.
     */
    onReject() { }
    #checkHalfOpen() {
        if (this.#state === 'open' && this.#clock() - this.#openedAt >= this.#resetTimeoutMs) {
            this.#state = 'halfOpen';
            this.#successCount = 0;
            this.onHalfOpen();
        }
    }
    #recordSuccess() {
        if (this.#state === 'halfOpen') {
            this.#successCount += 1;
            if (this.#successCount >= this.#successThreshold) {
                this.#state = 'closed';
                this.#failureCount = 0;
                this.#successCount = 0;
                this.#openedAt = 0;
                this.onSuccess();
                this.onClose();
            }
            else {
                this.onSuccess();
            }
        }
        else {
            this.#failureCount = 0;
            this.onSuccess();
        }
    }
    #recordFailure(err) {
        if (this.#state === 'halfOpen') {
            this.#state = 'open';
            this.#openedAt = this.#clock();
            this.onFailure(err);
            this.onOpen();
            return;
        }
        this.#failureCount += 1;
        if (this.#failureCount >= this.#failureThreshold) {
            this.#state = 'open';
            this.#openedAt = this.#clock();
            this.onFailure(err);
            this.onTrip();
            this.onOpen();
        }
        else {
            this.onFailure(err);
        }
    }
}
