// @studnicky/resilience/dist/DeadLetterQueue.js
/** Bounded FIFO DLQ with async-generator drain; enqueue() throws on capacity/closed/aborted. */
import { DeadLetterQueueBuilder } from './DeadLetterQueueBuilder.js';
import { DlqAbortedError } from './DlqAbortedError.js';
import { DlqClosedError } from './DlqClosedError.js';
import { DlqFullError } from './DlqFullError.js';
import { ResilienceConfigError } from './errors/ResilienceConfigError.js';
export class DeadLetterQueue {
    #capacity;
    #clock;
    #entries = [];
    #closed = false;
    #aborted = false;
    #notifyDrain = null;
    static builder() {
        const factory = (options) => {
            const result = DeadLetterQueue.create(options);
            return result;
        };
        const result = DeadLetterQueueBuilder.create(factory);
        return result;
    }
    static create(options) {
        return new DeadLetterQueue(options);
    }
    constructor(options) {
        const capacity = options?.capacity ?? Infinity;
        if (capacity !== undefined && (capacity <= 0 || Number.isNaN(capacity))) {
            throw new ResilienceConfigError('capacity must be > 0');
        }
        this.#capacity = capacity;
        this.#clock = options?.clock ?? (() => { const result = Date.now(); return result; });
        const signal = options?.signal;
        if (signal !== undefined) {
            if (signal.aborted) {
                this.#aborted = true;
            }
            else {
                signal.addEventListener('abort', () => { this.#abort(); }, { 'once': true });
            }
        }
    }
    get size() {
        const result = this.#entries.length;
        return result;
    }
    get closed() {
        const result = this.#closed;
        return result;
    }
    /** Throws DlqFullError | DlqClosedError | DlqAbortedError on failure. */
    enqueue(item, reason, error) {
        if (this.#aborted) {
            throw new DlqAbortedError();
        }
        if (this.#closed) {
            throw new DlqClosedError();
        }
        if (this.#entries.length >= this.#capacity) {
            this.onOverflow();
            throw new DlqFullError();
        }
        this.#entries.push({ 'enqueuedAtMs': this.#clock(), 'error': error, 'id': crypto.randomUUID(), 'item': item, 'reason': reason });
        if (this.#notifyDrain !== null) {
            const n = this.#notifyDrain;
            this.#notifyDrain = null;
            n();
        }
        this.onEnqueue(item);
    }
    async *drain() {
        while (true) {
            const entry = this.#entries.shift();
            if (entry !== undefined) {
                this.onDequeue(entry.item);
                yield entry;
                continue;
            }
            if (this.#closed || this.#aborted) {
                return;
            }
            await new Promise((resolve) => { this.#notifyDrain = resolve; });
        }
    }
    close() {
        this.#closed = true;
        if (this.#notifyDrain !== null) {
            const n = this.#notifyDrain;
            this.#notifyDrain = null;
            n();
        }
        this.onClose();
    }
    abort() { this.#abort(); }
    /**
     * Fires after an item is added to the queue.
     * Override to add logging, metrics, or tracing. Must not throw or block.
     */
    onEnqueue(_item) { }
    /**
     * Fires after an item is shifted from the queue during drain.
     * Override to add logging, metrics, or tracing. Must not throw or block.
     */
    onDequeue(_item) { }
    /**
     * Fires when `enqueue()` is called on a full queue, before throwing DlqFullError.
     * Must not throw or block.
     */
    onOverflow() { }
    /**
     * Fires at the end of `close()`.
     * Must not throw or block.
     */
    onClose() { }
    /**
     * Fires at the end of `#abort()`.
     * Must not throw or block.
     */
    onAbort() { }
    #abort() {
        this.#aborted = true;
        if (this.#notifyDrain !== null) {
            const n = this.#notifyDrain;
            this.#notifyDrain = null;
            n();
        }
        this.onAbort();
    }
}
