// @studnicky/resilience/dist/TokenBucket.js
/** Token bucket rate limiter; consume() throws when exhausted, waitForToken() blocks until available. */
import { ResilienceConfigError } from './errors/ResilienceConfigError.js';
import { TokenBucketBuilder } from './TokenBucketBuilder.js';
import { TokenBucketExhaustedError } from './TokenBucketExhaustedError.js';
export class TokenBucket {
    #requestsPerSecond;
    #burstSize;
    #clock;
    #tokens;
    #lastRefill;
    static builder() {
        const factory = (options) => {
            const result = TokenBucket.create(options);
            return result;
        };
        const result = TokenBucketBuilder.create(factory);
        return result;
    }
    static create(options) {
        return new this(options);
    }
    constructor(options) {
        if (options.requestsPerSecond <= 0) {
            throw new ResilienceConfigError('requestsPerSecond must be > 0');
        }
        if (options.burstSize < 1) {
            throw new ResilienceConfigError('burstSize must be >= 1');
        }
        this.#requestsPerSecond = options.requestsPerSecond;
        this.#burstSize = options.burstSize;
        this.#clock = options.clock ?? (() => { const result = Date.now(); return result; });
        this.#tokens = options.burstSize;
        this.#lastRefill = this.#clock();
    }
    get available() {
        this.#refill();
        return this.#tokens;
    }
    /** Throws TokenBucketExhaustedError if no token available. */
    consume(tokens = 1) {
        this.#refill();
        if (this.#tokens < tokens) {
            this.onTokenDepleted();
            throw new TokenBucketExhaustedError();
        }
        this.#tokens -= tokens;
        this.onTokenAcquired(tokens);
    }
    /** Wait until tokens are available, then consume. */
    async waitForToken(options = {}) {
        const tokens = options.tokens ?? 1;
        const signal = options.signal;
        while (true) {
            this.#refill();
            if (this.#tokens >= tokens) {
                this.#tokens -= tokens;
                this.onTokenAcquired(tokens);
                return;
            }
            const waitMs = Math.ceil((tokens - this.#tokens) / this.#requestsPerSecond * 1000);
            await new Promise((resolve, reject) => {
                const timer = setTimeout(resolve, waitMs);
                signal?.addEventListener('abort', () => { clearTimeout(timer); reject(signal.reason); }, { 'once': true });
            });
        }
    }
    /**
     * Fires after `consume()` or `waitForToken()` successfully deducts tokens.
     * Override to add logging, metrics, or tracing. Must not throw or block.
     */
    onTokenAcquired(_count) { }
    /**
     * Fires when `consume()` finds insufficient tokens, before throwing.
     * Must not throw or block.
     */
    onTokenDepleted() { }
    /**
     * Fires when the internal refill adds tokens due to elapsed time.
     * Only fires when `added > 0`. Must not throw or block.
     */
    onRefill(_added) { }
    #refill() {
        const now = this.#clock();
        const elapsed = now - this.#lastRefill;
        const newTokens = (elapsed / 1000) * this.#requestsPerSecond;
        const prev = this.#tokens;
        this.#tokens = Math.min(this.#burstSize, this.#tokens + newTokens);
        this.#lastRefill = now;
        const added = this.#tokens - prev;
        if (added > 0) {
            this.onRefill(added);
        }
    }
}
