// @studnicky/resilience/dist/TokenBucketBuilder.js
/** Fluent builder for TokenBucket. */
import { ResilienceConfigError } from './errors/ResilienceConfigError.js';
export class TokenBucketBuilder {
    #create;
    #requestsPerSecond;
    #burstSize;
    #clock;
    static create(create) {
        return new TokenBucketBuilder(create);
    }
    constructor(create) {
        this.#create = create;
    }
    withRequestsPerSecond(value) {
        this.#requestsPerSecond = value;
        return this;
    }
    withBurstSize(value) {
        this.#burstSize = value;
        return this;
    }
    withClock(value) {
        this.#clock = value;
        return this;
    }
    build() {
        if (this.#requestsPerSecond === undefined) {
            throw new ResilienceConfigError('requestsPerSecond is required');
        }
        if (this.#burstSize === undefined) {
            throw new ResilienceConfigError('burstSize is required');
        }
        const options = {
            'burstSize': this.#burstSize,
            'requestsPerSecond': this.#requestsPerSecond,
            ...(this.#clock !== undefined ? { 'clock': this.#clock } : {})
        };
        return this.#create(options);
    }
}
