// @studnicky/eslint-config/dist/rules/requireOptionsObject.js
const DEFAULT_MIN_OPTIONALS = 2;
class ParamInspector {
    static isOptional(param) {
        if (param.type === 'RestElement') {
            return false;
        }
        if (param.type === 'ObjectPattern') {
            return false;
        }
        if (param.type === 'AssignmentPattern') {
            return true;
        }
        if (param.type === 'Identifier') {
            const p = param;
            return p.optional === true;
        }
        return false;
    }
    static isOptionsObject(param) {
        if (param.type === 'AssignmentPattern') {
            const p = param;
            return p.left.type === 'ObjectPattern';
        }
        if (param.type === 'Identifier') {
            const p = param;
            const ann = p.typeAnnotation;
            if (ann === undefined || ann === null) {
                return false;
            }
            return ann.typeAnnotation.type === 'TSTypeLiteral';
        }
        return false;
    }
    static check(params, context, node, name, minOptionals) {
        const optionals = params.filter(ParamInspector.isOptional);
        if (optionals.length < minOptionals) {
            return;
        }
        const lastOptional = optionals[optionals.length - 1];
        if (lastOptional !== undefined && ParamInspector.isOptionsObject(lastOptional)) {
            return;
        }
        context.report({
            'data': { 'count': String(optionals.length), 'name': name },
            'messageId': 'requireOptionsObject',
            'node': node
        });
    }
}
class FunctionName {
    static fromParent(node) {
        const parent = node.parent;
        if (parent.type === 'VariableDeclarator' && parent.id?.type === 'Identifier') {
            return parent.id.name;
        }
        if ((parent.type === 'MethodDefinition' || parent.type === 'Property') && parent.key?.type === 'Identifier') {
            return parent.key.name;
        }
        return '(anonymous)';
    }
}
export const requireOptionsObject = {
    'create': (context) => {
        const rawOptions = context.options[0];
        const minOptionals = rawOptions?.minOptionals ?? DEFAULT_MIN_OPTIONALS;
        const listener = {
            'ArrowFunctionExpression': (node) => {
                const n = node;
                const name = FunctionName.fromParent(node);
                ParamInspector.check(n.params, context, node, name, minOptionals);
            },
            'FunctionDeclaration': (node) => {
                const n = node;
                const name = n.id?.name ?? '(anonymous)';
                ParamInspector.check(n.params, context, node, name, minOptionals);
            },
            'FunctionExpression': (node) => {
                const n = node;
                const name = FunctionName.fromParent(node);
                ParamInspector.check(n.params, context, node, name, minOptionals);
            }
        };
        const tsListener = listener;
        tsListener.TSCallSignatureDeclaration = (node) => {
            const n = node;
            ParamInspector.check(n.params, context, node, '(anonymous)', minOptionals);
        };
        tsListener.TSConstructSignatureDeclaration = (node) => {
            const n = node;
            ParamInspector.check(n.params, context, node, '(anonymous)', minOptionals);
        };
        tsListener.TSMethodSignature = (node) => {
            const n = node;
            const name = n.key?.type === 'Identifier' ? (n.key.name ?? '(anonymous)') : '(anonymous)';
            ParamInspector.check(n.params, context, node, name, minOptionals);
        };
        return listener;
    },
    'meta': {
        'docs': {
            'description': 'Require 2+ optional parameters to be collected into a single trailing options object.',
            'recommended': false
        },
        'messages': {
            'requireOptionsObject': "Callable '{{name}}' has {{count}} optional parameters. Collect them into a single trailing options object: '{{name}}(required, opts?: { optA?, optB? })'."
        },
        'schema': [
            {
                'additionalProperties': false,
                'properties': {
                    'minOptionals': {
                        'default': 2,
                        'description': 'Minimum number of optional parameters to trigger the rule.',
                        'minimum': 2,
                        'type': 'number'
                    }
                },
                'type': 'object'
            }
        ],
        'type': 'suggestion'
    }
};
