// @studnicky/errors/dist/errors/ModuleError.js
import { ErrorDefaults } from '../constants/index.js';
import { BaseError } from './BaseError.js';
import { ValidationError } from './ValidationError.js';
/**
 * Base error for all modules. Extends `BaseError` while preserving the
 * scenario-defaults API and `context: Record<string, unknown> | undefined`
 * typing from the original implementation.
 */
export class ModuleError extends BaseError {
    /**
     * Create a new ModuleError with scenario defaults.
     *
     * Merges user options over the specified scenario defaults from `ErrorDefaults`.
     * User-provided options take precedence over scenario defaults.
     */
    static create(message, options) {
        if (!(options.scenario in ErrorDefaults)) {
            throw ValidationError.create({
                'message': `Must be one of: ${Object.keys(ErrorDefaults).join(', ')}`,
                'path': 'scenario',
                'violations': [
                    {
                        'message': `Invalid error scenario: ${String(options.scenario)}`,
                        'path': 'scenario'
                    }
                ]
            });
        }
        const defaults = ErrorDefaults[options.scenario];
        const mergedOptions = {
            'cause': options.cause,
            'code': defaults.code,
            'context': options.context,
            'retryable': options.retryable ?? defaults.retryable,
            'statusCode': options.statusCode ?? defaults.statusCode
        };
        return new ModuleError(message, mergedOptions);
    }
    /**
     * Untyped context dictionary typed as `Record<string, unknown> | undefined`.
     *
     * Shadows `BaseError.metadata` for the `ModuleError` public API surface.
     */
    context;
    /**
     * Typed cause — narrows `Error.cause: unknown` to `Error | undefined`.
     */
    cause;
    /** HTTP status code (for API/HTTP errors). */
    statusCode;
    /**
     * Protected constructor — use `ModuleError.create()` instead.
     * Subclasses can call this constructor directly in their own `create()` methods.
     */
    constructor(message, options) {
        if (typeof message !== 'string' || message.length === 0) {
            throw ValidationError.create({
                'message': 'Must be a non-empty string',
                'path': 'message'
            });
        }
        if (typeof options.code !== 'string' || options.code.length === 0) {
            throw ValidationError.create({
                'message': 'Must be a non-empty string',
                'path': 'code'
            });
        }
        super({
            'cause': options.cause,
            'code': options.code,
            'message': message,
            'retryable': options.retryable ?? false
        });
        this.cause = options.cause;
        this.statusCode = options.statusCode;
        this.context = options.context;
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, this.constructor);
        }
    }
    /**
     * Walks the cause chain starting at `this`, returning all `Error` nodes.
     * Subclasses may override to alter traversal behavior.
     */
    walkCauseChain() {
        const chain = [this];
        let current = this.cause;
        while (current instanceof Error) {
            chain.push(current);
            current = current.cause;
        }
        return chain;
    }
    /**
     * Returns extra serialized fields to merge into the `toJSON()` output.
     * Override in subclasses to inject additional fields; call `super.serializeExtra()`
     * to preserve the standard ModuleError fields.
     *
     * Fire-point: called from `toJSON()`.
     * Default returns `{}`.
     */
    serializeExtra() {
        const extra = {};
        return extra;
    }
    /**
     * Finds the first cause of a specific type in the chain.
     */
    findCauseOfType(errorType) {
        const found = this.walkCauseChain().find((error) => { return error instanceof errorType; });
        return found;
    }
    /**
     * Returns the full error chain including all causes.
     */
    getCauseChain() {
        const result = this.walkCauseChain();
        return result;
    }
    /**
     * Returns `true` if this error or any cause is an instance of `errorType`.
     */
    hasCauseOfType(errorType) {
        const result = this.walkCauseChain().some((error) => { return error instanceof errorType; });
        return result;
    }
    /**
     * Serializes this error for structured logging.
     * Builds the ModuleError-format JSON object, then merges `serializeExtra()`
     * so subclasses can inject additional fields without rewriting cause-chain logic.
     */
    toJSON() {
        const json = {
            'code': this.code,
            'message': this.message,
            'name': this.name,
            'retryable': this.retryable,
            'stack': this.stack
        };
        if (this.statusCode !== undefined) {
            json.statusCode = this.statusCode;
        }
        if (this.context !== undefined) {
            json.context = this.context;
        }
        const cause = this.cause;
        if (cause !== undefined) {
            if (cause instanceof ModuleError) {
                json.cause = cause.toJSON();
            }
            else if (cause instanceof Error) {
                json.cause = {
                    'message': cause.message,
                    'name': cause.name,
                    'stack': cause.stack
                };
            }
            else {
                json.cause = cause;
            }
        }
        return { ...json, ...this.serializeExtra() };
    }
}
