// @studnicky/errors/dist/errors/ValidationError.js
import { BaseError } from './BaseError.js';
import { ErrorCodeRegistry } from './ErrorCodeRegistry.js';
// Register the error code at module load.
ErrorCodeRegistry.register({
    'code': 'errors.validationFailed',
    'description': 'Input validation failed against the declared schema.',
    'retryable': false
});
/**
 * Thrown when schema/type validation fails on a public API boundary.
 * Code: `'errors.validationFailed'`.
 *
 * Use `ValidationError.create()` as the ergonomic construction path.
 * Subclasses may call the protected constructor directly from their own
 * `create()` factory.
 */
export class ValidationError extends BaseError {
    /**
     * Creates a new `ValidationError`.
     */
    static create(args) {
        const result = new ValidationError(args);
        return result;
    }
    /** Structured list of validation violations. */
    violations;
    constructor(args) {
        const metadata = { 'path': args.path };
        // Don't try to serialize violations into JsonValueType — store them separately.
        super({
            'code': 'errors.validationFailed',
            'correlationId': args.correlationId,
            'message': ValidationError.buildMessage(args.path, args.message),
            'metadata': metadata,
            'retryable': false
        });
        this.violations = args.violations;
    }
    /**
     * Builds the formatted error message from path and summary.
     * Subclasses may override to customize the message template; they must pass
     * the result as the `message` to `super()` in their own constructor.
     *
     * Fire-point: called as a static helper from the constructor initializer.
     */
    static buildMessage(path, summary) {
        const result = `Validation failed at "${path}": ${summary}`;
        return result;
    }
    /**
     * Returns extra serialized fields for `toJSON()`.
     * Includes `violations` when defined.
     *
     * Fire-point: called from `toJSON()` via the base `serializeExtra()` hook.
     */
    serializeExtra() {
        if (this.violations !== undefined) {
            return { 'violations': this.violations };
        }
        return {};
    }
    /**
     * Formats the user-facing message.
     * Includes violation details when present.
     *
     * Fire-point: called from `toUserMessage()`.
     */
    formatUserMessage() {
        if (this.violations !== undefined && this.violations.length > 0) {
            const lines = this.violations.map((v) => { const result = `  - ${v.path}: ${v.message}`; return result; });
            return `${this.message}\n${lines.join('\n')}`;
        }
        return this.message;
    }
    toUserMessage() {
        const result = this.formatUserMessage();
        return result;
    }
}
