// @studnicky/errors/dist/errors/ValidationErrors.js
/** Iterable collection of validation violations with RFC 7807 reporting. */
import { ValidationError } from './ValidationError.js';
import { ValidationErrorsBuilder } from './ValidationErrorsBuilder.js';
const DEFAULT_PROBLEM_TYPE = 'https://problems.studnicky.dev/validation';
/**
 * Ordered, iterable collection of `ValidationViolationType` items.
 *
 * NOT a thrown error — returned by validators. Use `ValidationError` for single-violation boundary exceptions.
 *
 * Construct via `ValidationErrors.create(items)` or `ValidationErrors.builder().addViolation(v).build()`.
 */
export class ValidationErrors {
    /** Creates a `ValidationErrors` from an array of violations. */
    static create(items) {
        const result = new this(items);
        return result;
    }
    /** Returns a fluent builder for assembling a `ValidationErrors` incrementally. */
    static builder() {
        const result = ValidationErrorsBuilder.create((items) => {
            const instance = ValidationErrors.create(items);
            return instance;
        });
        return result;
    }
    /** Creates a `ValidationErrors` from an array of violations. */
    static of(violations) {
        const result = ValidationErrors.create(violations);
        return result;
    }
    /** Merges multiple `ValidationErrors` collections into one. */
    static merge(...errors) {
        const violations = errors.flatMap((e) => {
            const items = [...e.items];
            return items;
        });
        const result = ValidationErrors.create(violations);
        return result;
    }
    /** Maps Ajv-style validator errors into a `ValidationErrors` instance; empty when rawErrors is null/empty. */
    static fromValidatorErrors(rawErrors) {
        if (rawErrors === null || rawErrors === undefined || rawErrors.length === 0) {
            const result = ValidationErrors.create([]);
            return result;
        }
        const violations = rawErrors.map((raw) => {
            const violation = {
                'keyword': raw.keyword,
                'message': raw.message ?? raw.keyword,
                'path': raw.instancePath
            };
            return violation;
        });
        const result = ValidationErrors.create(violations);
        return result;
    }
    /** The raw ordered list of validation violations. */
    items;
    constructor(items) {
        if (!Array.isArray(items)) {
            throw ValidationError.create({ 'message': 'items must be an array', 'path': 'items' });
        }
        this.items = items;
    }
    /** Compact rollup of deduplicated, sorted paths and keywords; safe for metric labels. */
    aggregate() {
        const pathSet = new Set();
        const keywordSet = new Set();
        for (const item of this.items) {
            pathSet.add(item.path);
            keywordSet.add(item.keyword);
        }
        return {
            'count': this.items.length,
            'keywords': [...keywordSet].sort(),
            'paths': [...pathSet].sort()
        };
    }
    /** RFC 7807 Problem Details payload; defaults: type validation URI, title 'Validation failed', status 422. */
    report(options) {
        const count = this.items.length;
        const detail = count === 1 ? '1 validation error' : `${count} validation errors`;
        return {
            'detail': detail,
            'errors': [...this.items],
            'status': options?.status ?? 422,
            'title': options?.title ?? 'Validation failed',
            'type': options?.type ?? DEFAULT_PROBLEM_TYPE
        };
    }
    /** Number of violations in this collection. */
    get length() {
        const result = this.items.length;
        return result;
    }
    /** `true` when there are no violations. */
    get ok() {
        return this.items.length === 0;
    }
    [Symbol.iterator]() {
        const result = this.items[Symbol.iterator]();
        return result;
    }
}
