// @studnicky/scheduler/dist/scheduler/RealTimeScheduler.js
import { RealTimeSchedulerBuilder } from './RealTimeSchedulerBuilder.js';
/**
 * Unified `ScheduledTask` implementation for both one-shot and interval timers.
 * Cancellation is fully delegated to the provided callback, which handles
 * timer clearing, map removal, and the `onCancel` lifecycle hook.
 */
class RealTimeTask {
    atMs;
    id;
    #onCancelCallback;
    /**
     * Property write order: atMs, id, #onCancelCallback.
     */
    constructor(atMs, id, onCancelCallback) {
        this.atMs = atMs;
        this.id = id;
        this.#onCancelCallback = onCancelCallback;
    }
    /** Cancels the underlying timer via the cancel callback. No-op if already fired/cancelled. */
    cancel() {
        this.#onCancelCallback(this.id);
    }
}
/**
 * Real-time `SchedulerProvider` using `setTimeout` and `setInterval`.
 * Instantiate once per scheduler context; inject as a `SchedulerProviderType`.
 *
 * Subclass extension seams:
 * - `generateId()` — override to control task ID generation
 * - `createTimeout(fire, delayMs)` — override to substitute the timer backend
 * - `createInterval(fire, intervalMs)` — override to substitute the timer backend
 * - `clearTimer(handle, variant)` — override to control timer clearing
 * - Lifecycle hooks: `onSchedule`, `onFire`, `onFireError`, `onDrift`, `onMiss`,
 *   `onCancel`, `onCancelAll`, `onIdle`
 */
export class RealTimeScheduler {
    /** Map from task ID to active timer. */
    #timers;
    #idCounter;
    /**
     * Property write order: #timers, #idCounter.
     */
    constructor() {
        this.#timers = new Map();
        this.#idCounter = 0;
    }
    /** Creates a new `RealTimeScheduler` instance. */
    static create() {
        return new this();
    }
    /** Returns a `RealTimeSchedulerBuilder` pre-wired to create `RealTimeScheduler` instances. */
    static builder() {
        const result = RealTimeSchedulerBuilder.create(() => { const instance = RealTimeScheduler.create(); return instance; });
        return result;
    }
    /** Returns a unique task ID. Override to customise the ID format. */
    generateId() {
        this.#idCounter = this.#idCounter + 1;
        return `task-${this.#idCounter.toString()}`;
    }
    /** Creates a one-shot timer. Override to substitute the timer backend. */
    createTimeout(fire, delayMs) {
        const result = setTimeout(fire, delayMs);
        return result;
    }
    /** Creates a repeating timer. Override to substitute the timer backend. */
    createInterval(fire, intervalMs) {
        const result = setInterval(fire, intervalMs);
        return result;
    }
    /** Clears a timer handle. Override to control timer clearing. */
    clearTimer(handle, variant) {
        if (variant === 'interval') {
            clearInterval(handle);
        }
        else {
            clearTimeout(handle);
        }
    }
    /** Called after a task is registered via `scheduleAt` or `scheduleEvery`. */
    onSchedule(_id, _atMs, _variant) { return; }
    /** Called inside the timer callback, immediately before `fire` is invoked. */
    onFire(_id) { return; }
    /**
     * Called when a fired task's callback throws synchronously or returns a rejected Promise.
     * The error is still silently swallowed by the scheduler — this hook is the only
     * observability seam for task-level failures.
     */
    onFireError(_id, _error) { return; }
    /**
     * Called when a one-shot task fires later than its scheduled `atMs`.
     * Only meaningful for `scheduleAt` — `scheduleEvery` tasks are not tracked against a
     * wall-clock due time in the same way.
     *
     * @param id - Task ID
     * @param dueMs - Originally requested epoch-ms fire time
     * @param actualMs - `Date.now()` at the moment of actual fire
     * @param driftMs - `actualMs - dueMs` (always positive when this hook fires)
     */
    onDrift(_id, _dueMs, _actualMs, _driftMs) { return; }
    /**
     * Called when `scheduleAt` receives an `atMs` that is already in the past
     * (`atMs < Date.now()` at scheduling time). The task will still be scheduled
     * to fire at the next event-loop turn.
     *
     * @param id - Task ID
     * @param atMs - The requested (past) fire time
     * @param nowMs - `Date.now()` at the time `scheduleAt` was called
     */
    onMiss(_id, _atMs, _nowMs) { return; }
    /** Called when a task's `cancel()` method is invoked. */
    onCancel(_id) { return; }
    /** Called at the end of `cancelAll()`, after all timers have been cleared. */
    onCancelAll() { return; }
    /**
     * Called after `cancelAll` drains all tracked tasks (no tasks remain active).
     * Not fired on individual `cancel()` calls — only when the scheduler is fully drained.
     */
    onIdle() { return; }
    /** Cancels and removes all currently-tracked tasks. */
    cancelAll() {
        for (const timer of this.#timers.values()) {
            this.clearTimer(timer.handle, timer.variant);
        }
        this.#timers.clear();
        this.onCancelAll();
        this.onIdle();
    }
    /**
     * Schedules `fire` once at `atMs` (epoch-ms).
     * If `atMs` is in the past, fires at the next event-loop turn (and `onMiss` fires).
     */
    scheduleAt(atMs, fire) {
        const id = this.generateId();
        const scheduleNowMs = Date.now();
        const rawDelayMs = atMs - scheduleNowMs;
        const delayMs = rawDelayMs < 0 ? 0 : rawDelayMs;
        const timers = this.#timers;
        if (rawDelayMs < 0) {
            this.onMiss(id, atMs, scheduleNowMs);
        }
        const handle = this.createTimeout(() => {
            timers.delete(id);
            this.onFire(id);
            const fireNowMs = Date.now();
            const drift = fireNowMs - atMs;
            if (drift > 0) {
                this.onDrift(id, atMs, fireNowMs, drift);
            }
            let result;
            try {
                result = fire();
            }
            catch (error) {
                this.onFireError(id, error);
                return;
            }
            if (result instanceof Promise) {
                result.catch((error) => {
                    this.onFireError(id, error);
                });
            }
        }, delayMs);
        const activeTimer = {
            'handle': handle,
            'id': id,
            'variant': 'timeout'
        };
        timers.set(id, activeTimer);
        const task = new RealTimeTask(atMs, id, (taskId) => {
            const timer = timers.get(taskId);
            if (timer === undefined) {
                return;
            }
            this.clearTimer(timer.handle, timer.variant);
            timers.delete(taskId);
            this.onCancel(taskId);
        });
        this.onSchedule(id, atMs, 'timeout');
        return task;
    }
    /**
     * Schedules `fire` to run repeatedly every `intervalMs` milliseconds.
     */
    scheduleEvery(intervalMs, fire) {
        const id = this.generateId();
        const atMs = Date.now() + intervalMs;
        const timers = this.#timers;
        const handle = this.createInterval(() => {
            this.onFire(id);
            let result;
            try {
                result = fire();
            }
            catch (error) {
                this.onFireError(id, error);
                return;
            }
            if (result instanceof Promise) {
                result.catch((error) => {
                    this.onFireError(id, error);
                });
            }
        }, intervalMs);
        const activeTimer = {
            'handle': handle,
            'id': id,
            'variant': 'interval'
        };
        timers.set(id, activeTimer);
        const task = new RealTimeTask(atMs, id, (taskId) => {
            const timer = timers.get(taskId);
            if (timer === undefined) {
                return;
            }
            this.clearTimer(timer.handle, timer.variant);
            timers.delete(taskId);
            this.onCancel(taskId);
        });
        this.onSchedule(id, atMs, 'interval');
        return task;
    }
}
