// @studnicky/system/dist/System.js
import { SystemProvider } from './providers/SystemProvider.js';
const PROVIDER = new SystemProvider();
export class System {
    static #gpuCache = undefined;
    constructor() {
        throw new Error('System is a static-only class');
    }
    static get cpu() {
        const logicalCount = PROVIDER.logicalCpuCount();
        const model = PROVIDER.cpuModel();
        const arch = PROVIDER.arch();
        const hasHyperthreading = arch !== 'arm64';
        const physicalCount = hasHyperthreading
            ? Math.max(1, Math.round(logicalCount / 2))
            : logicalCount;
        return {
            'arch': arch,
            'logicalCount': logicalCount,
            'model': model,
            'physicalCount': physicalCount
        };
    }
    static get memory() {
        return {
            'freeMb': PROVIDER.freeMb(),
            'totalMb': PROVIDER.totalMb()
        };
    }
    static get platform() {
        const platformStr = PROVIDER.platform();
        const arch = PROVIDER.arch();
        return {
            'isAppleSilicon': platformStr === 'darwin' && arch === 'arm64',
            'nodeVersion': PROVIDER.runtimeVersion(),
            'os': platformStr
        };
    }
    static gpu() {
        if (System.#gpuCache !== undefined) {
            return System.#gpuCache;
        }
        const detected = PROVIDER.detectGpu();
        System.#gpuCache = detected;
        return System.#gpuCache;
    }
    static get logicalCpuCount() {
        const result = PROVIDER.logicalCpuCount();
        return result;
    }
    static get optimalWorkerCount() {
        const result = Math.max(1, PROVIDER.logicalCpuCount() - 1);
        return result;
    }
    static get isAppleSilicon() {
        const platformStr = PROVIDER.platform();
        const arch = PROVIDER.arch();
        return platformStr === 'darwin' && arch === 'arm64';
    }
    static snapshot() {
        const gpu = System.gpu();
        return {
            'cpu': System.cpu,
            'gpu': gpu,
            'memory': System.memory,
            'platform': System.platform
        };
    }
}
