// @studnicky/json/dist/json/Clone.js
/**
 * Clone — deep cloning with Map/Set/Date/Array support.
 *
 * Deep-clones plain objects, arrays, Map, Set and Date; other values (including
 * class instances) are returned as-is.
 *
 * Subclass `Clone` and override any `protected static clone*` step to customise
 * cloning behaviour.
 */
export class Clone {
    // ---------------------------------------------------------------------------
    // Protected steps — override in subclasses to customise cloning
    // ---------------------------------------------------------------------------
    /** Clone a single value by dispatching to the appropriate protected step. */
    static cloneValue(value) {
        if (value === null || typeof value !== 'object') {
            return value;
        }
        if (Array.isArray(value)) {
            return this.cloneArray(value);
        }
        if (value instanceof Map) {
            return this.cloneMap(value);
        }
        if (value instanceof Set) {
            return this.cloneSet(value);
        }
        if (value instanceof Date) {
            return this.cloneDate(value);
        }
        return this.cloneObject(value);
    }
    /** Clone an array element-by-element. */
    static cloneArray(value) {
        const result = value.map((item) => { const result = this.cloneValue(item); return result; });
        return result;
    }
    /** Clone a Map, deep-cloning both keys and values. */
    static cloneMap(value) {
        const cloned = new Map();
        for (const [k, v] of value.entries()) {
            cloned.set(this.cloneValue(k), this.cloneValue(v));
        }
        return cloned;
    }
    /** Clone a Set, deep-cloning each member. */
    static cloneSet(value) {
        const cloned = new Set();
        for (const v of value.values()) {
            cloned.add(this.cloneValue(v));
        }
        return cloned;
    }
    /** Clone a Date by timestamp. */
    static cloneDate(value) {
        const result = new Date(value.getTime());
        return result;
    }
    /**
     * Clone a plain/class object by own enumerable keys.
     * Override to customise per-object behaviour (e.g. tag injection).
     */
    static cloneObject(value) {
        const cloned = {};
        const keys = Object.keys(value);
        const keysLen = keys.length;
        for (let i = 0; i < keysLen; i += 1) {
            const key = keys[i];
            cloned[key] = this.cloneValue(value[key]);
        }
        return cloned;
    }
    // ---------------------------------------------------------------------------
    // Public static API
    // ---------------------------------------------------------------------------
    /**
     * Recursively deep-clone a value.
     *
     * Handles: primitives, arrays, Map, Set, Date, and plain objects.
     * Class instances with custom prototypes are shallow-property-cloned
     * (own enumerable keys only) — use `structuredClone` when you need
     * transfer semantics for non-plain classes.
     */
    static deep(value) {
        const result = this.cloneValue(value);
        return result;
    }
    /**
     * Shallow clone a plain-object record.
     */
    static shallow(value) {
        const clone = { ...value };
        return clone;
    }
}
