// @studnicky/concurrency/dist/Semaphore.js
/** Counting permit gate. acquire() returns a release function. */
import { SemaphoreOptionsEntity } from './entities/SemaphoreOptionsEntity.js';
import { SemaphoreError } from './errors/SemaphoreError.js';
import { SemaphoreBuilder } from './SemaphoreBuilder.js';
export class Semaphore {
    static builder() {
        const result = SemaphoreBuilder.create((options) => {
            const semaphore = Semaphore.create(options);
            return semaphore;
        });
        return result;
    }
    static create(options) {
        const result = new this(options);
        return result;
    }
    static #validate(options) {
        if (!SemaphoreOptionsEntity.validate(options)) {
            throw new SemaphoreError('permits must be a positive integer');
        }
        if (!Number.isInteger(options.permits) || options.permits < 1) {
            throw new SemaphoreError('permits must be a positive integer');
        }
    }
    #available;
    #permits;
    #queue;
    constructor(options) {
        Semaphore.#validate(options);
        this.#available = options.permits;
        this.#permits = options.permits;
        this.#queue = [];
    }
    get available() {
        const result = this.#available;
        return result;
    }
    get permits() {
        const result = this.#permits;
        return result;
    }
    async acquire() {
        if (this.#available > 0) {
            const permitsBefore = this.#available;
            this.#available -= 1;
            this.onAcquire(permitsBefore);
            return this.#buildRelease();
        }
        this.onAcquireWait();
        return await new Promise((resolve) => {
            this.#queue.push(() => { resolve(this.#buildRelease()); });
            this.onContended(this.#queue.length);
        });
    }
    async withPermit(callback) {
        const release = await this.acquire();
        try {
            return await callback();
        }
        finally {
            release();
        }
    }
    #buildRelease() {
        let released = false;
        return () => {
            if (released) {
                return;
            }
            released = true;
            this.#release();
        };
    }
    #release() {
        const next = this.#queue.shift();
        if (next !== undefined) {
            next();
            this.onReleaseDelegated();
            return;
        }
        this.#available += 1;
        this.onRelease(this.#available);
    }
    /**
     * Fires when a permit is granted immediately.
     * `permitsBefore` is the available count BEFORE decrement.
     * Overrides must not throw or block.
     */
    onAcquire(_permitsBefore) { }
    /**
     * Fires when the caller had to queue (no permit available).
     * Overrides must not throw or block.
     */
    onAcquireWait() { }
    /**
     * Fires when a new waiter is added to the queue.
     * `queueLength` is the queue length AFTER push.
     * Overrides must not throw or block.
     */
    onContended(_queueLength) { }
    /**
     * Fires when a permit is returned to the pool (no waiting callers).
     * `permitsAfter` is the available count AFTER increment.
     * Overrides must not throw or block.
     */
    onRelease(_permitsAfter) { }
    /**
     * Fires when a permit is handed to a queued waiter (not returned to pool).
     * Overrides must not throw or block.
     */
    onReleaseDelegated() { }
}
